<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Fast single trait Genome Wide Association Studies (GWAS) following 
    the method described in Kang et al. (2010), &lt;doi:10.1038/ng.548&gt;.        
    One of a series of statistical genetic packages for streamlining the 
    analysis of typical plant breeding experiments developed by Biometris.">
<title>An R package for fast GWAS â€¢ statgenGWAS</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="An R package for fast GWAS">
<meta property="og:description" content="Fast single trait Genome Wide Association Studies (GWAS) following 
    the method described in Kang et al. (2010), &lt;doi:10.1038/ng.548&gt;.        
    One of a series of statistical genetic packages for streamlining the 
    analysis of typical plant breeding experiments developed by Biometris.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">statgenGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.9.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="articles/index.html">Vignettes</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    functions
  </a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown--other-packages">
    <span class="fa fa-seedling"></span>
     
    other packages
  </a>
  <div class="dropdown-menu" aria-labelledby="dropdown--other-packages">
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenSTA">statgenSTA</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenGxE">statgenGxE</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenHTP">statgenHTP</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenIBD">statgenIBD</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenMPP">statgenMPP</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/LMMsolver">LMMsolver</a>
  </div>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Biometris/statgenGWAS">
    <span class="fa fa-github fa-lg fab"></span>
     
    github
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="statgengwas">statgenGWAS<a class="anchor" aria-label="anchor" href="#statgengwas"></a>
</h1></div>
<p><a href="https://www.r-pkg.org/pkg/statgenGWAS" class="external-link"><img src="https://www.r-pkg.org/badges/version/statgenGWAS"></a> <a href="https://www.r-pkg.org/pkg/statgenGWAS" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/statgenGWAS" alt="CRAN RStudio mirror downloads"></a> <a href="https://github.com/Biometris/statgenGWAS/actions?workflow=R-CMD-check" class="external-link"><img src="https://github.com/Biometris/statgenGWAS/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a> <a href="https://app.codecov.io/gh/Biometris/statgenGWAS" class="external-link"><img src="https://codecov.io/gh/Biometris/statgenGWAS/branch/master/graph/badge.svg" alt="codecov"></a></p>
<p><strong>statgenGWAS</strong> is an R package for fast single trait Genome Wide Association Studies (GWAS).</p>
<p>statgenGWAS has extensive options for summarizing and visualizing results. The example below shows a selection of what is possible. For a full overview of all options it is best to read the <a href="https://biometris.github.io/statgenGWAS/articles/GWAS.html" class="external-link"><strong>vignette</strong></a>.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<ul>
<li>Install from CRAN:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"statgenGWAS"</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Install latest development version from GitHub (requires <a href="https://github.com/r-lib/remotes" class="external-link">remotes</a> package):</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"Biometris/statgenGWAS"</span>, ref <span class="op">=</span> <span class="st">"develop"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<p>Example using the data from the European Union project DROPS. The same data is used in the vignette and described there in detail.</p>
<p>First get the data in the form required by the package, converting it to an object of class <strong>g</strong>(enomic)<strong>Data</strong>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://biometris.github.io/statgenGWAS/index.html" class="external-link">statgenGWAS</a></span><span class="op">)</span></span>
<span><span class="co">## Read data.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dropsMarkers"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dropsMap"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dropsPheno"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Add genotypes as row names of dropsMarkers and drop Ind column.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dropsMarkers</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">dropsMarkers</span><span class="op">[[</span><span class="st">"Ind"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">dropsMarkers</span> <span class="op">&lt;-</span> <span class="va">dropsMarkers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsMarkers</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"Ind"</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Add genotypes as row names of dropsMap.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dropsMap</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">dropsMap</span><span class="op">[[</span><span class="st">"SNP.names"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## Rename Chomosome and Position columns.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsMap</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Chromosome"</span>, <span class="st">"Position"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsMap</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">"pos"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Rename Variety_ID in phenotypic data to genotype.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsPheno</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsPheno</span><span class="op">)</span> <span class="op">==</span> <span class="st">"Variety_ID"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"genotype"</span></span>
<span><span class="co">## Select relevant columns and convert data to a list.</span></span>
<span><span class="va">dropsPhenoList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dropsPheno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"genotype"</span>, <span class="st">"grain.yield"</span>,</span>
<span>                                         <span class="st">"grain.number"</span>, <span class="st">"seed.size"</span>,</span>
<span>                                         <span class="st">"anthesis"</span>, <span class="st">"silking"</span>, <span class="st">"plant.height"</span>,</span>
<span>                                         <span class="st">"tassel.height"</span>, <span class="st">"ear.height"</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                        f <span class="op">=</span> <span class="va">dropsPheno</span><span class="op">[[</span><span class="st">"Experiment"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Create a gData object all data.</span></span>
<span><span class="va">gDataDrops</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/gData.html">createGData</a></span><span class="op">(</span>geno <span class="op">=</span> <span class="va">dropsMarkers</span>, map <span class="op">=</span> <span class="va">dropsMap</span>, pheno <span class="op">=</span> <span class="va">dropsPhenoList</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Remove duplicate SNPs from gDataDrops.</span></span>
<span><span class="va">gDataDropsDedup</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/codeMarkers.html">codeMarkers</a></span><span class="op">(</span><span class="va">gDataDrops</span>, impute <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Input contains 41722 SNPs for 246 genotypes.</span></span>
<span><span class="co">#&gt; 0 genotypes removed because proportion of missing values larger than or equal to 1.</span></span>
<span><span class="co">#&gt; 0 SNPs removed because proportion of missing values larger than or equal to 1.</span></span>
<span><span class="co">#&gt; 5098 duplicate SNPs removed.</span></span>
<span><span class="co">#&gt; Output contains 36624 SNPs for 246 genotypes.</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run single trait GWAS for traits 'grain.yield' and 'anthesis' for trial Mur13W.</span></span>
<span><span class="va">GWASDrops</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/runSingleTraitGwas.html">runSingleTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grain.yield"</span>, <span class="st">"anthesis"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## QQ plot of GWAS Drops.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"qq"</span>, trait <span class="op">=</span> <span class="st">"grain.yield"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-qqStg-1.png" width="75%"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Manhattan plot of GWAS Drops.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"manhattan"</span>, trait <span class="op">=</span> <span class="st">"grain.yield"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-manhattanStg-1.png" width="75%"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Qtl plot of GWAS Drops.</span></span>
<span><span class="co">## Set significance threshold to 4 and normalize effect estimates.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"qtl"</span>, yThr <span class="op">=</span> <span class="fl">4</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-qtlStgNorm-1.png" width="75%"></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=statgenGWAS" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/Biometris/statgenGWAS/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Biometris/statgenGWAS/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing statgenGWAS</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Bart-Jan van Rossum <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-8673-2514" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Willem Kruijer <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-7179-1733" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Bart-Jan van Rossum, Willem Kruijer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
