<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • statgenGWAS</title><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><link href="../deps/KaTex-0.16.10/katex.min.css" rel="stylesheet"><script src="../deps/KaTex-0.16.10/katex.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">statgenGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.11</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Vignettes</a></li>
      </ul><ul class="navbar-nav"><li class="active nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> news</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-file-code-o"></span> functions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-other-packages" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-seedling"></span> other packages</button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-other-packages"><li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenSTA">statgenSTA</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenGxE">statgenGxE</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenHTP">statgenHTP</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenIBD">statgenIBD</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenMPP">statgenMPP</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/statgenQTLxT">statgenQTLxT</a></li>
    <li><a class="external-link dropdown-item" href="https://biometris.github.io/LMMsolver">LMMsolver</a></li>
  </ul></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Biometris/statgenGWAS"><span class="fa fa-github fa-lg fab"></span> github</a></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Changelog</h1>
      <small>Source: <a href="https://github.com/Biometris/statgenGWAS/blob/main/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.11" id="statgengwas-1011">statgenGWAS 1.0.11<a class="anchor" aria-label="anchor" href="#statgengwas-1011"></a></h2>
<ul><li>Random imputation is now done properly for marker matrices with values larger than 2.</li>
<li>Estimating variance components using the NR algorithm no also works when replicated genotypes are used.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.10" id="statgengwas-1010">statgenGWAS 1.0.10<a class="anchor" aria-label="anchor" href="#statgengwas-1010"></a></h2><p class="text-muted">CRAN release: 2024-11-15</p>
<ul><li>The QTL plot now has a legend explaining the meaning of the colors used.</li>
<li>The support for tibbles and data.tables with phenotypic data is improved.</li>
<li>Functions no longer rely on soft-deprecated ggplot2 functions.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.9" id="statgengwas-109">statgenGWAS 1.0.9<a class="anchor" aria-label="anchor" href="#statgengwas-109"></a></h2><p class="text-muted">CRAN release: 2022-10-13</p>
<ul><li>The kinship function now has an argument MAF allowing for excluding SNPs with a low minor allele frequency from the calculations.</li>
<li>Small change in tests to accommodate future ggplot2 release.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.8" id="statgengwas-108">statgenGWAS 1.0.8<a class="anchor" aria-label="anchor" href="#statgengwas-108"></a></h2><p class="text-muted">CRAN release: 2022-02-16</p>
<ul><li>A function readPLINK is added that reads file in PLINK format and converts them directly to a gData object.</li>
<li>The kinship function now has an option MAF allowing for excluding markers with a MAF below a specified threshold.</li>
<li>The gData object now has a plot function allowing plotting of the genetic map.</li>
<li>A bug in gData summaries when only one trait is present in the phenotypic data is fixed.</li>
<li>A bug in runSingleTraitGwas is fixed now allowing for using traits that are in pheno as covariates in covar.</li>
<li>Sorting of chromosomes in map is now done correctly for numeric values stored as character.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.7" id="statgengwas-107">statgenGWAS 1.0.7<a class="anchor" aria-label="anchor" href="#statgengwas-107"></a></h2><p class="text-muted">CRAN release: 2021-07-29</p>
<ul><li>No user visual changes.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.6" id="statgengwas-106">statgenGWAS 1.0.6<a class="anchor" aria-label="anchor" href="#statgengwas-106"></a></h2><p class="text-muted">CRAN release: 2021-07-26</p>
<ul><li>The plot function now has an option title.</li>
<li>The codeMarkers function now has an option MAC equivalent to the same option in runSingleTraitGwas.</li>
<li>Bug when using tibbles instead of data.frames as input for covariates is fixed.</li>
<li>The kinship function now has an option for returning a named identity matrix as output.</li>
<li>An extra check for the presence of . in a character marker matrix is added.</li>
<li>The correct SNPs are now excluded when using MAF/MAC.</li>
<li>The function documentation and vignette have been revised and extra clarification is added where needed.</li>
<li>The beagle software used for imputation is updated to version 5.2.</li>
<li>Bug in the computation of the genomic inflation factor is fixed. It was incorrectly computed as 1 / genomic inflation factor.</li>
<li>The data in the vignette is updated such that the environments included match those in the other statgen packages.</li>
<li>The manhattan plot function gained arguments for setting the start and end position when plotting results for a single chromosome.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.5" id="statgengwas-105">statgenGWAS 1.0.5<a class="anchor" aria-label="anchor" href="#statgengwas-105"></a></h2><p class="text-muted">CRAN release: 2020-09-29</p>
<ul><li>An option fdr (false discovery rate) is added for selecting significant SNPs when running single trait GWAS. The procedure for this is based on Brzyski et al. (2017). See the function documentation and vignette for an extensive explanation.</li>
<li>The summary function for objects of class GWAS now has an option traits for restricting the number of traits for which the summary is printed.</li>
<li>Problem where codeMarkers was getting very slow for large character matrix inputs is fixed by moving part of the code to c++.</li>
<li>Full ggplot2 is no longer imported.</li>
<li>Removed use of deprecated <code><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">ggplot2::expand_scale</a></code>.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.4" id="statgengwas-104">statgenGWAS 1.0.4<a class="anchor" aria-label="anchor" href="#statgengwas-104"></a></h2><p class="text-muted">CRAN release: 2020-03-02</p>
<ul><li>stringsAsFactors = TRUE added where applicable to comply with new defaults in R 4.0.</li>
<li>Bug in plotting of GWAS objects fixed. The first trait was always plotted in case more than one trait was present.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.3" id="statgengwas-103">statgenGWAS 1.0.3<a class="anchor" aria-label="anchor" href="#statgengwas-103"></a></h2><p class="text-muted">CRAN release: 2020-01-27</p>
<ul><li>Dependency on deprecated <code>rvg::ph_with_vg_at</code> removed.</li>
<li>OMP_THREAD_LIMIT is now respected in parallel code.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.2" id="statgengwas-102">statgenGWAS 1.0.2<a class="anchor" aria-label="anchor" href="#statgengwas-102"></a></h2><p class="text-muted">CRAN release: 2020-01-20</p>
<ul><li>Fixed problem with compilation on Solaris</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.1" id="statgengwas-101">statgenGWAS 1.0.1<a class="anchor" aria-label="anchor" href="#statgengwas-101"></a></h2><p class="text-muted">CRAN release: 2020-01-19</p>
<ul><li>Initial CRAN version</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bart-Jan van Rossum, Willem Kruijer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

