---
title: "Introduction to the statgenGWAS package"
author: "Bart-Jan van Rossum"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    toc: false
    number_sections: true
vignette: >
  %\VignetteIndexEntry{Introduction to the statgenGWAS package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
  
```{r setup, include = FALSE}
knitr::opts_chunk$set(
collapse = TRUE,
comment = "#>",
fig.dim = c(7, 4)
)
library(statgenGWAS)
```


The statgenGWAS package has been designed for performing Genome Wide Association Studies (GWAS).

****

## The class `gData`

Data for analysis on genomic data comes from different sources and is stored in one data object of class **gData** (**g**enomic **Data**) for convenience. A gData object will contain all data needed for performing analyses, so the first thing to do when using the statgenGWAS package is creating a gData object. A gData object consists of different types of data. To create the object all data is optional, however for certain analyses certain types of data are required. Additional data can always be added to an existing object later on.  
A gData object can be created using the function **_createGData_**. If an existing gData object is provided as an input to this function, data is added to the existing object and existing data will be overwritten with a warning if necessary. All data is checked for consistency and, if needed, default names are added with a warning message.  
A gData object created this way will consist of the following components:  

### **Marker map**  
The data.frame **map** is used to describe the physical positions of the markers on the chromosomes. The data consists of two columns, *chr* for the name or number of the chromosome and *pos* for the position of the marker on the chromosome. The position can be in basepair or in centimorgan. The names of the markers should be the row names of the data.frame.  

### **Marker matrix**  
The marker matrix is stored in the matrix **marker** within the gData object. It has the names of the markers in its column names and the genotypes in its row names. Markers can either be coded as character strings or as numerical values. In the former case, before performing any analysis, the marker matrix has to be converted to a numerical matrix. This can be do using the function **_codeMarkers_**. 

### **Phenotypic data**  
Phenotypic data, either directly from field trials or after summarizing can be stored in **pheno** in the gData object. Pheno can either be a single data.frame or a list of data.frames for storing data for different trials or possibly different summarizations of the original data. The first column of all elements of pheno should be *genotype* and all the other columns should represent different traits. Storing additional variables should be done in **covar**. A single data.frame will be converted to a one-item list upon creation of a gData object. 

### **Kinship matrix**  
The kinship matrix describes the genetic relatedness between the different genotypes. It is stored in the matrix **kinship** in the gData object. This should be a square matrix with genotypes in both row and column names and a measure for the genetic relatedness in its cells. There are many ways of computing genetic relatedness some of which are included in this package. Instead of providing one kinship matrix a list of kinship matrices can be provided with each matrix in this list a chromosome specific matrix, usually computed using all markers that are not on the corresponding chromosome. If a named list of kinship matrices is provided, its names should match the names or numbers of the chromosomes in **map**. Unnamed lists will be given default names. 

### **Further covariates**  
Further covariates that can be used in the analyses can be stored in the data.frame **covar**. This data.frame has genotypes in its row names and the covariates in the column names. Covariates can be numerical, factor or character data. The latter will be converted to a factor when creating the gData object.

****

## Single Trait GWAS

stG

****

## Worked example

As an example of the functionality of the package a worked example is provided using maize data from the European Union project DROPS. The data is available from https://data.inra.fr/dataset.xhtml?persistentId=doi:10.15454/IASSTN and the relevant data sets are included as data.frames in the statgenGWAS package.

### **Load data**

The first step is loading the data into R. 
```{r loadData} 
data(dropsMarkers)
data(dropsMap)
data(dropsPheno)
```
dropsMarkers contains the coded marker information for 41722 SNPs and 246 genotypes. dropsMap contains information about the positions of those SNPs on the B73 reference genome V2. dropsPheno contains data for the genotypic means (Best Linear Unbiased Estimation, BLUEs), with one value per experiment per genotype, for eight traits. For a more detailed description of the contents of the data see ```help(dropsData)```

### **Create gData object**

Using the raw data the next step is to create a **gData** object that can be used for all analyses. For this the raw data has to be converted to a suitable format for a **gData** object.
When creating a **gData** object the markers should be input as a matrix or data.frame with genotypes in the row names and markers in the column names. The latter is already satisfied so only the row names should be added and the Ind column used for that should be dropped.
```{r convertMarkers}
## Add genotypes as row names of dropsMarkers and drop Ind column.
rownames(dropsMarkers) <- dropsMarkers$Ind
dropsMarkers <- dropsMarkers[colnames(dropsMarkers) != "Ind"]
```

The map used as input when creating a **gData** object should at least have the columns "chr" and "pos" and the row names should be the names of the SNPs. Other columns are ignored when creating a **gData** object so they can be left as they are.
```{r convertMap}
## Add genotypes as row names of dropsMap.
rownames(dropsMap) <- dropsMap$SNP.names
## Rename Chomosome and Position columns.
colnames(dropsMap)[match(c("Chromosome", "Position"), colnames(dropsMap))] <-
  c("chr", "pos")
```

Now a gData object containing only map and marker information can be created. This object can be used as a base and phenotypic data, a kinship matrix and covariates may be added later on.
```{r createGdata}
## Create a gData object containing map and marker information.
gDataDrops <- createGData(geno = dropsMarkers, map = dropsMap)
```

The next step involves adding phenotypic data to the gData object just created. dropsPheno contains genotypic means for 27 trials. To be able to run a GWAS analysis per trial the data has to be added as a list of 27 data.frames. These data.frames should have "genotype" as their first column and may only contain traits after that. Other columns need to be dropped.    
The phenotypic data can be added to *gDataDrops* using the same function that was used for creating the object. Adding an argument *gData = * to the function call will add new info to an existing object. If gDataDrops would have been non-empty that the phenotypic data already present would be overwritten.
```{r addPheno}
## Convert phenotypic data to a list
dropsPhenoList <- split(x = dropsPheno, f = dropsPheno$Experiment)
## Rename Variety_ID to genotype and select relevant columns.
dropsPhenoList <- lapply(X = dropsPhenoList, FUN = function(trial) {
  colnames(trial)[colnames(trial) == "Variety_ID"] <- "genotype"
  trial <- trial[c("genotype", "grain.yield", "grain.number", "seed.size",
                   "anthesis", "silking", "plant.height", "tassel.height",
                   "ear.height")]
  return(trial)
})
## Add phenotypic data to gDataDrops
gDataDrops <- createGData(gData = gDataDrops, pheno = dropsPhenoList)
```
Note that if the phenotypic data consists of only one trial/experiment it can be added as a single data.frame without first converting it to a list. In this case createGData will still convert the input to a list with one item.

To get an idea of the contents of the data a summary of the gData object can be made. This will give an overview of the content of the map and markers and also print a summary per trait per trial. Since there are 29 trials and eight traits in gDataDrops the output is restricted here to one trial using the `trials` parameter of the summary functions
```{r sumGData}
## Summarize gDataDrops
summary(gDataDrops, trials = "Bol12R")
```

### **Coding of markers**

Marker data has to be numerical and without missing values in order to do GWAS analysis. The marker data available for the DROPS project has already been converted from A/T/C/G to 0/1/2. However, still the *codeMarkers* function in the package can be used to remove duplicate markers.
```{r removeDupMarkers}
## Remove duplicate markers from gDataDrops
gDataDropsDedup <- codeMarkers(gDataDrops, impute = FALSE, verbose = TRUE) 
```

Note that duplicate markers are removed at random. To replicate the result of previous runs make sure to set a seed.

Therefore before running the actual analysis the markers in *gDataRice* need to be coded and inputed. For this the package contains a function *codeMarkers*. Running this function with the default settings will recode the markers using the minor allele in a SNP as reference allele for this SNP. After that imputation will be done with a random value based on the distribution of values within a single SNP. 

```{r eval=FALSE}
gDataRiceCoded <- codeMarkers(gData = gDataRice)
```

Imputation with a random value isn't the best way to impute the markers. For a better imputation the independent beagle software can be used. This software is included in the package and will be used when calling *codeMarkers* with the proper settings. Doing so will call beagle with its default settings. For the current example beagle will take a long time to run. It is advised to load the results from the beagle imputation that come with the package directly.

```{r eval=FALSE}
## Perform imputation using beagle.
gDataRiceCodedBeagle <- codeMarkers(gdata = gDataRice, 
                                    removeDuplicates = FALSE, 
                                    imputeType = "beagle")
```

More options in codeMarkers take care of removing duplicate SNPs, removing SNPs with a low Minor Allele Frequency (MAF) and removing SNPs with a high percentage of missing value. See the function documentation for an overview of these options.

### **Single trait GWAS**

The cleaned gData object can be used for performing single trait GWAS analysis. With all the default settings the analysis already performs quite well.
```{r stg}
## Run single trait GWAS for trial 'Bol12W' and trait 'grain.yield'
GWASDrops <- runSingleTraitGwas(gData = gDataDropsDedup, traits = "grain.yield", trials = "Bol12W")
```
To inspect the output a summary and plots the can be made. The default plot is a manhattan plot, other types of plot can be specified using the *plotType* argument in the function.
```{r sumStg}
summary(GWASDrops)
```

```{r qqStg} 
plot(GWASDrops, plotType = "qq")
```

```{r manhattanStg}
plot(GWASDrops, plotType = "manhattan")
```


In the default function call of the *runSingleTraitGwas* function a single kinship matrix is computed using the algorithm proposed by Astle and Balding. In most cases computing chromosome specific kinship matrices considerably improves the predicting power of the algorithm. Several different methods for computing the kinship matrix are available in the package. Users can provide their own kinship matrices as well.  
As a default significant SNP are selected with a threshold computed using Bonferroni correction. Other thresholds can be specified as well. More options can be found in the function documentation for *runSingleTraitGwas*.
The following example performs the GWAS computation using chromosome specific kinship matrices computed using Identity By State (IBS), a threshold for significant SNPs of 4 and PC1 - PC4 as covariates.
```{r eval=FALSE}
GWASRicePlantHeight <- 
  runSingleTraitGwas(gData = gDataRiceCodedBeagle, traits = "Plant.height", 
                     covar = c("PC1", "PC2", "PC3", "PC4"), GLSMethod = "multi",
                     kinshipMethod = "IBS", thrType = "fixed", LODThr = 4)

summary(GWASRicePlantHeight)
plot(GWASRicePlantHeight, type = "qq")
plot(GWASRicePlantHeight)
```
The analysis can be performed for more than one trait in a single function call. Not specifying traits at all will perform the analysis for all traits that are present in the phenotypic data. The results will be combined in a single output object with the traits specified in the column *trait* in *GWAResult* and *signSnp* within this object.  
When making a qq or manhattan plot of this output the trait has to be specified. Also a "qtl" plot is available which plots significant SNPs per trait against their position on the chromosomes. 
```{r eval=FALSE}
GWASRicePlantHeightMult <- 
  runSingleTraitGwas(gData = gDataRiceCodedBeagle, 
                     covar = c("PC1", "PC2", "PC3", "PC4"), GLSMethod = "multi",
                     kinshipMethod = "IBS", thrType = "fixed", LODThr = 4)

summary(GWASRicePlantHeightMult)
plot(GWASRicePlantHeightMult, type = "qq", trait = "Plant.height")
plot(GWASRicePlantHeightMult, type = "qtl")
```

****
