<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Bart-Jan van Rossum" />

<meta name="date" content="2019-11-20" />

<title>Introduction to the statgenGWAS package</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to the statgenGWAS package</h1>
<h4 class="author">Bart-Jan van Rossum</h4>
<h4 class="date">2019-11-20</h4>



<p>The statgenGWAS package has been designed for performing single trait Genome Wide Association Studies (GWAS).</p>
<hr />
<div id="the-class-gdata" class="section level1">
<h1><span class="header-section-number">1</span> The class <code>gData</code></h1>
<p>Data for analysis on genomic data comes from different sources and is stored in one data object of class <strong>gData</strong> (<strong>g</strong>enomic <strong>Data</strong>) for convenience. A gData object will contain all data needed for performing analyses, so the first thing to do when using the statgenGWAS package is creating a gData object. A gData object consists of different types of data. To create the object all data is optional, however for certain analyses certain types of data are required. Additional data can always be added to an existing object later on.<br />
A gData object can be created using the function <strong><em>createGData</em></strong>. If an existing gData object is provided as an input to this function, data is added to the existing object and existing data will be overwritten with a warning if necessary. All data is checked for consistency and, if needed, default names are added with a warning message.<br />
A gData object created this way will consist of the following components:</p>
<div id="marker-map" class="section level2">
<h2><span class="header-section-number">1.1</span> <strong>Marker map</strong></h2>
<p>The data.frame <strong>map</strong> is used to describe the physical positions of the markers on the chromosomes. The data consists of two columns, <em>chr</em> for the name or number of the chromosome and <em>pos</em> for the position of the marker on the chromosome. The position can be in basepair or in centimorgan. The names of the markers should be the row names of the data.frame.</p>
</div>
<div id="marker-matrix" class="section level2">
<h2><span class="header-section-number">1.2</span> <strong>Marker matrix</strong></h2>
<p>The marker matrix is stored in the matrix <strong>marker</strong> within the gData object. It has the names of the markers in its column names and the genotypes in its row names. Markers can either be coded as character strings or as numerical values. In the former case, before performing any analysis, the marker matrix has to be converted to a numerical matrix. This can be do using the function <strong><em>codeMarkers</em></strong>.</p>
</div>
<div id="phenotypic-data" class="section level2">
<h2><span class="header-section-number">1.3</span> <strong>Phenotypic data</strong></h2>
<p>Phenotypic data, either directly from field trials or after summarizing can be stored in <strong>pheno</strong> in the gData object. Pheno can either be a single data.frame or a list of data.frames for storing data for different trials or possibly different summarizations of the original data. The first column of all elements of pheno should be <em>genotype</em> and all the other columns should represent different traits. Storing additional variables should be done in <strong>covar</strong>. A single data.frame will be converted to a one-item list upon creation of a gData object.</p>
</div>
<div id="kinship-matrix" class="section level2">
<h2><span class="header-section-number">1.4</span> <strong>Kinship matrix</strong></h2>
<p>The kinship matrix describes the genetic relatedness between the different genotypes. It is stored in the matrix <strong>kinship</strong> in the gData object. This should be a square matrix with genotypes in both row and column names and a measure for the genetic relatedness in its cells. There are many ways of computing genetic relatedness some of which are included in this package. Instead of providing one kinship matrix a list of kinship matrices can be provided with each matrix in this list a chromosome specific matrix, usually computed using all markers that are not on the corresponding chromosome. If a named list of kinship matrices is provided, its names should match the names or numbers of the chromosomes in <strong>map</strong>. Unnamed lists will be given default names.</p>
</div>
<div id="further-covariates" class="section level2">
<h2><span class="header-section-number">1.5</span> <strong>Further covariates</strong></h2>
<p>Further covariates that can be used in the analyses can be stored in the data.frame <strong>covar</strong>. This data.frame has genotypes in its row names and the covariates in the column names. Covariates can be numerical, factor or character data. The latter will be converted to a factor when creating the gData object.</p>
<hr />
</div>
</div>
<div id="single-trait-gwas" class="section level1">
<h1><span class="header-section-number">2</span> Single Trait GWAS</h1>
<p>stG</p>
<hr />
</div>
<div id="worked-example" class="section level1">
<h1><span class="header-section-number">3</span> Worked example</h1>
<p>As an example of the functionality of the package a worked example is provided using maize data from the European Union project DROPS. The data is available from <a href="https://data.inra.fr/dataset.xhtml?persistentId=doi:10.15454/IASSTN" class="uri">https://data.inra.fr/dataset.xhtml?persistentId=doi:10.15454/IASSTN</a> and the relevant data sets are included as data.frames in the statgenGWAS package.</p>
<div id="load-data" class="section level2">
<h2><span class="header-section-number">3.1</span> <strong>Load data</strong></h2>
<p>The first step is loading the data into R.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">data</span>(dropsMarkers)</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw">data</span>(dropsMap)</a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw">data</span>(dropsPheno)</a></code></pre></div>
<p>dropsMarkers contains the coded marker information for 41722 SNPs and 246 genotypes. dropsMap contains information about the positions of those SNPs on the B73 reference genome V2. dropsPheno contains data for the genotypic means (Best Linear Unbiased Estimation, BLUEs), with one value per experiment per genotype, for eight traits. For a more detailed description of the contents of the data see <code>help(dropsData)</code></p>
</div>
<div id="create-gdata-object" class="section level2">
<h2><span class="header-section-number">3.2</span> <strong>Create gData object</strong></h2>
<p>Using the raw data the next step is to create a <strong>gData</strong> object that can be used for all analyses. For this the raw data has to be converted to a suitable format for a <strong>gData</strong> object. When creating a <strong>gData</strong> object the markers should be input as a matrix or data.frame with genotypes in the row names and markers in the column names. The latter is already satisfied so only the row names should be added and the Ind column used for that should be dropped.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="co">## Add genotypes as row names of dropsMarkers and drop Ind column.</span></a>
<a class="sourceLine" id="cb2-2" title="2"><span class="kw">rownames</span>(dropsMarkers) &lt;-<span class="st"> </span>dropsMarkers<span class="op">$</span>Ind</a>
<a class="sourceLine" id="cb2-3" title="3">dropsMarkers &lt;-<span class="st"> </span>dropsMarkers[<span class="kw">colnames</span>(dropsMarkers) <span class="op">!=</span><span class="st"> &quot;Ind&quot;</span>]</a></code></pre></div>
<p>The map used as input when creating a <strong>gData</strong> object should at least have the columns “chr” and “pos” and the row names should be the names of the SNPs. Other columns are ignored when creating a <strong>gData</strong> object so they can be left as they are.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co">## Add genotypes as row names of dropsMap.</span></a>
<a class="sourceLine" id="cb3-2" title="2"><span class="kw">rownames</span>(dropsMap) &lt;-<span class="st"> </span>dropsMap<span class="op">$</span>SNP.names</a>
<a class="sourceLine" id="cb3-3" title="3"><span class="co">## Rename Chomosome and Position columns.</span></a>
<a class="sourceLine" id="cb3-4" title="4"><span class="kw">colnames</span>(dropsMap)[<span class="kw">match</span>(<span class="kw">c</span>(<span class="st">&quot;Chromosome&quot;</span>, <span class="st">&quot;Position&quot;</span>), <span class="kw">colnames</span>(dropsMap))] &lt;-</a>
<a class="sourceLine" id="cb3-5" title="5"><span class="st">  </span><span class="kw">c</span>(<span class="st">&quot;chr&quot;</span>, <span class="st">&quot;pos&quot;</span>)</a></code></pre></div>
<p>Now a gData object containing only map and marker information can be created. This object can be used as a base and phenotypic data, a kinship matrix and covariates may be added later on.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="co">## Create a gData object containing map and marker information.</span></a>
<a class="sourceLine" id="cb4-2" title="2">gDataDrops &lt;-<span class="st"> </span><span class="kw">createGData</span>(<span class="dt">geno =</span> dropsMarkers, <span class="dt">map =</span> dropsMap)</a></code></pre></div>
<p>The next step involves adding phenotypic data to the gData object just created. dropsPheno contains genotypic means for 27 trials. To be able to run a GWAS analysis per trial the data has to be added as a list of 27 data.frames. These data.frames should have “genotype” as their first column and may only contain traits after that. Other columns need to be dropped.<br />
The phenotypic data can be added to <em>gDataDrops</em> using the same function that was used for creating the object. Adding an argument <em>gData = </em> to the function call will add new info to an existing object. If gDataDrops would have been non-empty that the phenotypic data already present would be overwritten.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="co">## Convert phenotypic data to a list</span></a>
<a class="sourceLine" id="cb5-2" title="2">dropsPhenoList &lt;-<span class="st"> </span><span class="kw">split</span>(<span class="dt">x =</span> dropsPheno, <span class="dt">f =</span> dropsPheno<span class="op">$</span>Experiment)</a>
<a class="sourceLine" id="cb5-3" title="3"><span class="co">## Rename Variety_ID to genotype and select relevant columns.</span></a>
<a class="sourceLine" id="cb5-4" title="4">dropsPhenoList &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dt">X =</span> dropsPhenoList, <span class="dt">FUN =</span> <span class="cf">function</span>(trial) {</a>
<a class="sourceLine" id="cb5-5" title="5">  <span class="kw">colnames</span>(trial)[<span class="kw">colnames</span>(trial) <span class="op">==</span><span class="st"> &quot;Variety_ID&quot;</span>] &lt;-<span class="st"> &quot;genotype&quot;</span></a>
<a class="sourceLine" id="cb5-6" title="6">  trial &lt;-<span class="st"> </span>trial[<span class="kw">c</span>(<span class="st">&quot;genotype&quot;</span>, <span class="st">&quot;grain.yield&quot;</span>, <span class="st">&quot;grain.number&quot;</span>, <span class="st">&quot;seed.size&quot;</span>,</a>
<a class="sourceLine" id="cb5-7" title="7">                   <span class="st">&quot;anthesis&quot;</span>, <span class="st">&quot;silking&quot;</span>, <span class="st">&quot;plant.height&quot;</span>, <span class="st">&quot;tassel.height&quot;</span>,</a>
<a class="sourceLine" id="cb5-8" title="8">                   <span class="st">&quot;ear.height&quot;</span>)]</a>
<a class="sourceLine" id="cb5-9" title="9">  <span class="kw">return</span>(trial)</a>
<a class="sourceLine" id="cb5-10" title="10">})</a>
<a class="sourceLine" id="cb5-11" title="11"><span class="co">## Add phenotypic data to gDataDrops</span></a>
<a class="sourceLine" id="cb5-12" title="12">gDataDrops &lt;-<span class="st"> </span><span class="kw">createGData</span>(<span class="dt">gData =</span> gDataDrops, <span class="dt">pheno =</span> dropsPhenoList)</a></code></pre></div>
<p>Note that if the phenotypic data consists of only one trial/experiment it can be added as a single data.frame without first converting it to a list. In this case createGData will still convert the input to a list with one item.</p>
<p>To get an idea of the contents of the data a summary of the gData object can be made. This will give an overview of the content of the map and markers and also print a summary per trait per trial. Since there are 29 trials and eight traits in gDataDrops the output is restricted here to one trial using the <code>trials</code> parameter of the summary functions</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="co">## Summarize gDataDrops</span></a>
<a class="sourceLine" id="cb6-2" title="2"><span class="kw">summary</span>(gDataDrops, <span class="dt">trials =</span> <span class="st">&quot;Bol12R&quot;</span>)</a>
<a class="sourceLine" id="cb6-3" title="3"><span class="co">#&gt; map</span></a>
<a class="sourceLine" id="cb6-4" title="4"><span class="co">#&gt;  Number of markers: 41722 </span></a>
<a class="sourceLine" id="cb6-5" title="5"><span class="co">#&gt;  Number of chromosomes: 10 </span></a>
<a class="sourceLine" id="cb6-6" title="6"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-7" title="7"><span class="co">#&gt; markers</span></a>
<a class="sourceLine" id="cb6-8" title="8"><span class="co">#&gt;  Number of markers: 41722 </span></a>
<a class="sourceLine" id="cb6-9" title="9"><span class="co">#&gt;  Number of genotypes: 246 </span></a>
<a class="sourceLine" id="cb6-10" title="10"><span class="co">#&gt;  Content:</span></a>
<a class="sourceLine" id="cb6-11" title="11"><span class="co">#&gt;   0 1 2 NA </span></a>
<a class="sourceLine" id="cb6-12" title="12"><span class="co">#&gt;   0.28 0.01 0.71 0 </span></a>
<a class="sourceLine" id="cb6-13" title="13"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-14" title="14"><span class="co">#&gt; pheno</span></a>
<a class="sourceLine" id="cb6-15" title="15"><span class="co">#&gt;  Number of trials: 37 </span></a>
<a class="sourceLine" id="cb6-16" title="16"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-17" title="17"><span class="co">#&gt;  Bol12R:</span></a>
<a class="sourceLine" id="cb6-18" title="18"><span class="co">#&gt;      Number of traits: 8 </span></a>
<a class="sourceLine" id="cb6-19" title="19"><span class="co">#&gt;      Number of genotypes: 246 </span></a>
<a class="sourceLine" id="cb6-20" title="20"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-21" title="21"><span class="co">#&gt;   grain.yield      grain.number    seed.size        anthesis    </span></a>
<a class="sourceLine" id="cb6-22" title="22"><span class="co">#&gt;  Min.   :0.4789   Min.   : 247   Min.   :155.8   Min.   :58.47  </span></a>
<a class="sourceLine" id="cb6-23" title="23"><span class="co">#&gt;  1st Qu.:2.0756   1st Qu.:1015   1st Qu.:188.4   1st Qu.:65.46  </span></a>
<a class="sourceLine" id="cb6-24" title="24"><span class="co">#&gt;  Median :2.6574   Median :1336   Median :201.5   Median :67.61  </span></a>
<a class="sourceLine" id="cb6-25" title="25"><span class="co">#&gt;  Mean   :2.6635   Mean   :1337   Mean   :202.2   Mean   :67.78  </span></a>
<a class="sourceLine" id="cb6-26" title="26"><span class="co">#&gt;  3rd Qu.:3.2154   3rd Qu.:1629   3rd Qu.:216.3   3rd Qu.:70.10  </span></a>
<a class="sourceLine" id="cb6-27" title="27"><span class="co">#&gt;  Max.   :4.7351   Max.   :2399   Max.   :276.7   Max.   :77.29  </span></a>
<a class="sourceLine" id="cb6-28" title="28"><span class="co">#&gt;     silking       plant.height   tassel.height     ear.height    </span></a>
<a class="sourceLine" id="cb6-29" title="29"><span class="co">#&gt;  Min.   :64.77   Min.   :161.7   Min.   :217.1   Min.   : 81.25  </span></a>
<a class="sourceLine" id="cb6-30" title="30"><span class="co">#&gt;  1st Qu.:71.14   1st Qu.:185.4   1st Qu.:237.4   1st Qu.: 94.45  </span></a>
<a class="sourceLine" id="cb6-31" title="31"><span class="co">#&gt;  Median :74.26   Median :191.1   Median :244.4   Median : 99.08  </span></a>
<a class="sourceLine" id="cb6-32" title="32"><span class="co">#&gt;  Mean   :74.47   Mean   :191.1   Mean   :244.5   Mean   : 99.73  </span></a>
<a class="sourceLine" id="cb6-33" title="33"><span class="co">#&gt;  3rd Qu.:77.52   3rd Qu.:197.5   3rd Qu.:250.9   3rd Qu.:104.35  </span></a>
<a class="sourceLine" id="cb6-34" title="34"><span class="co">#&gt;  Max.   :85.65   Max.   :217.4   Max.   :286.2   Max.   :125.50</span></a></code></pre></div>
</div>
<div id="recoding-and-cleaning-of-markers" class="section level2">
<h2><span class="header-section-number">3.3</span> <strong>Recoding and cleaning of markers</strong></h2>
<p>Marker data has to be numerical and without missing values in order to do GWAS analysis. The marker data available for the DROPS project has already been converted from A/T/C/G to 0/1/2. However, still the <em>codeMarkers</em> function in the package can be used to remove duplicate markers.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="co">## Remove duplicate markers from gDataDrops</span></a>
<a class="sourceLine" id="cb7-2" title="2">gDataDropsDedup &lt;-<span class="st"> </span><span class="kw">codeMarkers</span>(gDataDrops, <span class="dt">impute =</span> <span class="ot">FALSE</span>, <span class="dt">verbose =</span> <span class="ot">TRUE</span>) </a>
<a class="sourceLine" id="cb7-3" title="3"><span class="co">#&gt; Input contains 41722 markers for 246 genotypes.</span></a>
<a class="sourceLine" id="cb7-4" title="4"><span class="co">#&gt; 0 genotypes removed because proportion of missing values larger than or equal to 1.</span></a>
<a class="sourceLine" id="cb7-5" title="5"><span class="co">#&gt; 0 markers removed because proportion of missing values larger than or equal to 1.</span></a>
<a class="sourceLine" id="cb7-6" title="6"><span class="co">#&gt; 5098 duplicate markers removed.</span></a>
<a class="sourceLine" id="cb7-7" title="7"><span class="co">#&gt; Output contains 36624 markers for 246 genotypes.</span></a></code></pre></div>
<p>Note that duplicate markers are removed at random. To replicate the result of previous runs make sure to set a seed.</p>
<p>To demonstrate the options of the <em>codeMarkers</em> function, we randomly replace 1% of the values in the marker matrix by NA.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co">## Copy gData object.</span></a>
<a class="sourceLine" id="cb8-2" title="2">gDataDropsMiss &lt;-<span class="st"> </span>gDataDrops</a>
<a class="sourceLine" id="cb8-3" title="3"><span class="co">## Add random missing values to 1% of the values in the marker matrix.</span></a>
<a class="sourceLine" id="cb8-4" title="4"><span class="kw">set.seed</span>(<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb8-5" title="5">nVal &lt;-<span class="st"> </span><span class="kw">nrow</span>(gDataDropsMiss<span class="op">$</span>markers) <span class="op">*</span><span class="st"> </span><span class="kw">ncol</span>(gDataDropsMiss<span class="op">$</span>markers)</a>
<a class="sourceLine" id="cb8-6" title="6">gDataDropsMiss<span class="op">$</span>markers[<span class="kw">sample</span>(<span class="dt">x =</span> <span class="dv">1</span><span class="op">:</span>nVal, <span class="dt">size =</span> nVal <span class="op">/</span><span class="st"> </span><span class="dv">100</span>)] &lt;-<span class="st"> </span><span class="ot">NA</span></a></code></pre></div>
<p>To re-impute the missing values added above, <em>codeMarkers</em> has to be run with the parameter impute set to TRUE. Then the type of imputation can be chosen setting the parameter imputeType:</p>
<ul>
<li>imputeType = “fixed”: Impute all missing values by a single fixed value set using the parameter fixedValue.</li>
<li>imputeType = “random”: Impute missing values with a random value based on the non-missing values for a SNP.</li>
<li>imputeType = “beagle”: Impute missing values using the independent beagle software <span class="citation">(Browning and Browning 2016)</span>. This software is include in the package and is called with its default settings. Since running beagle takes a long time for the current data the call to beagle is only shown here but not run.</li>
</ul>
<p>Before imputation genotypes with a high proportion of missing values in the marker matrix can be removed by setting the parameter nMissGeno. Setting it to 0.1 will remove all genotypes with a proportion of missing values higher than 0.1. In a similar way SNPs with a high proportion of missing values can be removed setting nMiss. The default value for both parameters is 1, so genotypes and SNPs with only missing values are always removed.<br />
The parameter MAF can be used to remove SNPs with a Minor Allele Frequency (MAF) lower than the specified MAF. As with removing duplicate SNPs this will be done both before and after imputation.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="co">## Impute missing values with random value.</span></a>
<a class="sourceLine" id="cb9-2" title="2"><span class="co">## Remove SNPs and genotypes with proportion of NA larger than 0.01</span></a>
<a class="sourceLine" id="cb9-3" title="3">gDataDropsImputed &lt;-<span class="st"> </span><span class="kw">codeMarkers</span>(<span class="dt">gData =</span> gDataDropsMiss, <span class="dt">nMissGeno =</span> <span class="fl">0.01</span>, </a>
<a class="sourceLine" id="cb9-4" title="4">                                 <span class="dt">nMiss =</span> <span class="fl">0.01</span>, <span class="dt">impute =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb9-5" title="5">                                 <span class="dt">imputeType =</span> <span class="st">&quot;random&quot;</span>, <span class="dt">verbose =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb9-6" title="6"><span class="co">#&gt; Input contains 41722 markers for 246 genotypes.</span></a>
<a class="sourceLine" id="cb9-7" title="7"><span class="co">#&gt; 119 genotypes removed because proportion of missing values larger than or equal to 0.01.</span></a>
<a class="sourceLine" id="cb9-8" title="8"><span class="co">#&gt; 14341 markers removed because proportion of missing values larger than or equal to 0.01.</span></a>
<a class="sourceLine" id="cb9-9" title="9"><span class="co">#&gt; 901 duplicate markers removed.</span></a>
<a class="sourceLine" id="cb9-10" title="10"><span class="co">#&gt; 15212 missing values imputed.</span></a>
<a class="sourceLine" id="cb9-11" title="11"><span class="co">#&gt; 1028 duplicate markers removed after imputation.</span></a>
<a class="sourceLine" id="cb9-12" title="12"><span class="co">#&gt; Output contains 25452 markers for 127 genotypes.</span></a></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co">## Impute missing values using beagle software.</span></a>
<a class="sourceLine" id="cb10-2" title="2">gDataDropsImputedBeagle &lt;-<span class="st"> </span><span class="kw">codeMarkers</span>(<span class="dt">gData =</span> gDataDropsMiss, <span class="dt">impute =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb10-3" title="3">                                       <span class="dt">imputeType =</span> <span class="st">&quot;beagle&quot;</span>, <span class="dt">verbose =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="single-trait-gwas-1" class="section level2">
<h2><span class="header-section-number">3.4</span> <strong>Single trait GWAS</strong></h2>
<p>The cleaned gData object can be used for performing single trait GWAS analysis. With all the default settings the analysis already performs quite well.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co">## Run single trait GWAS for trial &#39;Bol12W&#39; and trait &#39;grain.yield&#39;</span></a>
<a class="sourceLine" id="cb11-2" title="2">GWASDrops &lt;-<span class="st"> </span><span class="kw">runSingleTraitGwas</span>(<span class="dt">gData =</span> gDataDropsDedup, <span class="dt">traits =</span> <span class="st">&quot;grain.yield&quot;</span>,                                 <span class="dt">trials =</span> <span class="st">&quot;Bol12W&quot;</span>)</a></code></pre></div>
<p>For a quick overview of the results, e.g. the number of significant SNPs, use the summary function.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1"><span class="co">## Create summary of GWASDrops.</span></a>
<a class="sourceLine" id="cb12-2" title="2"><span class="kw">summary</span>(GWASDrops)</a>
<a class="sourceLine" id="cb12-3" title="3"><span class="co">#&gt; Bol12W:</span></a>
<a class="sourceLine" id="cb12-4" title="4"><span class="co">#&gt;  Traits analysed: grain.yield </span></a>
<a class="sourceLine" id="cb12-5" title="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb12-6" title="6"><span class="co">#&gt;  Data are available for 36624 SNPs.</span></a>
<a class="sourceLine" id="cb12-7" title="7"><span class="co">#&gt;   0 of them were not analyzed because their minor allele frequency is below 0.01 </span></a>
<a class="sourceLine" id="cb12-8" title="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb12-9" title="9"><span class="co">#&gt;  Trait: grain.yield </span></a>
<a class="sourceLine" id="cb12-10" title="10"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb12-11" title="11"><span class="co">#&gt;      Mixed model with only polygenic effects, and no marker effects:</span></a>
<a class="sourceLine" id="cb12-12" title="12"><span class="co">#&gt;      Genetic variance: 0.4262415 </span></a>
<a class="sourceLine" id="cb12-13" title="13"><span class="co">#&gt;      Residual variance: 0.3856793 </span></a>
<a class="sourceLine" id="cb12-14" title="14"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb12-15" title="15"><span class="co">#&gt;      LOD-threshold: 5.864796 </span></a>
<a class="sourceLine" id="cb12-16" title="16"><span class="co">#&gt;      No significant SNPs found. </span></a>
<a class="sourceLine" id="cb12-17" title="17"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb12-18" title="18"><span class="co">#&gt;      No genomic control correction was applied</span></a>
<a class="sourceLine" id="cb12-19" title="19"><span class="co">#&gt;      Genomic control inflation-factor: 0.972</span></a></code></pre></div>
<p>A qq plot of observed LOD scores against expected LOD scores <span class="citation">(Segura et al. 2012)</span> can be made by setting <code>plotType = &quot;qq&quot;</code> when plotting the results.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"><span class="co">## Plot a qq plot of GWAS Drops.</span></a>
<a class="sourceLine" id="cb13-2" title="2"><span class="kw">plot</span>(GWASDrops, <span class="dt">plotType =</span> <span class="st">&quot;qq&quot;</span>)</a></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqAAAAGACAMAAABFpiBcAAABMlBMVEUAAAAAADoAAGYAAP8AOjoAOmYAOpAAZrYzMzM6AAA6ADo6AGY6OgA6OpA6ZmY6ZpA6ZrY6kLY6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmAGZmOgBmOmZmOpBmZmZmZrZmkJBmkLZmkNtmtttmtv9uTU1uTW5uTY5ubo5ubqtuq+SOTU2OTW6OTY6Obk2OyP+QOgCQOjqQOmaQZgCQZjqQZpCQkGaQkLaQtpCQttuQtv+Q2/+rbk2rbm6rbo6ryKur5OSr5P+2ZgC2Zjq2Zma2kDq2kGa2kJC2tra2ttu22/+2/7a2///Ijk3I///bkDrbkGbbtpDbtrbb29vb2//b/7bb///kq27k///r6+v/tmb/tpD/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T////tfYxpAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAWzUlEQVR4nO2dC3sctRWGxzHEMdCbN4B7vzkUAiUubaGtoRBTKKVJlpY2jo2X2lnv//8LHc3s7M7laHSZkXSk+b6Hh3idfS1p8loaaWa02QpBGCcLXQEE6QsERVgHgiKsA0ER1oGgCOtAUIR1ICjCOhAUYR0IOkaWn7+SZS/88in1qvnG49v1by9+8shH9WIOBB0h3/44K7Jz0n3VTEvQ+S0IqggEHZ7lcfa9L3Mx/5AJ35qvOm+FoGaBoMMzzw5qXzRfFbk+PLjYz148qQQVfeyLXwqVs817EDoQdHCWx1U/eH14+2nzVfXVd/fLMb8Q9GK/PAOAoBqBoINzfbhXfXl661HzVfWObO/p8rNsrxA09/LXon/NVcUQrwwEHRwdQYu+NH8tBF3sF+84zTtUCKoMBB2cupK3nzZfzcVgfrD+3lyM6kLQI/HqAoLqBIIOTv2sc6/1CoIODQQdngsx01kev/jlXCx9Nl8VKYd4oS6GeNNA0OEpVz6//X05JW++KpJPkn5AT5KoxXykHgg6Qq7X146yYi7UfFX8/eELYmUp7y4by0zCUiwzKQJBx8j66vtnYvm9/WpVnI3mUorrS+VC/aJcqC9c3qMu2SObQNAx8+0fn1KvajN7xDAQ1EMgqH0gqIdAUPtAUA+BoPaBoAjrQFCEdSAowjoQFGEdCIqwDgRFWMdI0POe9P5lb0CC7JIQFCRrEoKCZE1CUJCsSQgKkjUJQUGyJiEoSNYkBAXJmoSgIFmTEBQkaxKCgmRNQlCQrEkICjI0mWWZnISgIAOTxeYWUhKCggxMQlCQnMkMgoJkSxZq4hwUJE+yLSZFQlCQYchOv0mTEBRkCFJuZ5PchaAgvZM9nWeT3N1FDwrSN6mwc0vmeuoP8Wez2ezVJxAU5DBS1XluyUJPfUEfP0APCnIoqWNnSa711Bb05qOHEBTkIFKr8yyyu9FTW9Dnb+dDfNGJ3smj6G0RpJNMe5azu0t9V4FfvfGw1oua/N7oB2SypH7nWYzttuugm/NQzWoZBmSipL6d61NPCArSG2nQeW5mRhaCXr72bHXzVywzgTQjjezczIxsetCz2ezuZiKvqpZdQCZG2nSenTJNhvhNeqtl2AyQiZIGdjb1hKAgnZOi89QmW3pCUJCOybLz1CQ7ekJQkI7I5v3xOuQuoScEBemGbD1dpCZJO88hKEg3ZOvpNxUp0xOCgnRBZu3HM/tJuZ4QFOTY5NrM5sJnH9mnJwQFOSopW4+XkvTMSEJCUJBDSM1nM2tR2XkOQUGORPZfyyRJDT0hKMhB5OZ8U3EtkyhTS08ICnII2Zmta5OaekJQkCZkU0ZtPdtlKmdGEhKCguzN2sdsY6aVoAZ2nkNQkAbpdpm6t3luyzTTE4KCNIh2hykt01RPCArSINZ+rss01xOCgtRPNqhMk5lRk9x+DUFB9mSIoHZ2no8gKDKVDNCC3iDEOuhBQXaTWZNiZ0878hxDPEjNZLZkubOnVZnn5xAUpF5MHn2rpZoZQVCQTsnMiqT3BzELBAWpjsGzmdvUJ+4QFKRDslqdNyGb60oQFKRD0lzQ9rInBAXpiKxf3tQkiWtGEBSkGzIzFlS5P4hZICjIvpgKKrmkCUFBOiEzM0GlV9whKMhRyMrG7dZfBjss9dwQAkFBjkFW/aXkYaO+MvvvpoOgIIeRdTGlD8PJyzTZH8QsEBTkedPMc+mzHbIy1Td7QlCQg8iamdVr7TJ17kWGoCAHkXqPw1Gk3q3yEBTkINJ6fxDNJzkgKMgBpO6zmi3Scn8Qs0DQqZGtJU6TR4kbZRo9BgdBQWomI2JRpuFTmhAUpGZGEdT4IWIIClIzwwW12YHBt6A3H1afxg1BYyO756BGZQ7fH8Se1Bf0bAZBIyUNPjeTKNN2gxDPgl69+S4EjZEsukv7Mq33r/Es6M1Hn5RD/J08yt4WYZPihNOaHnn/mgFRteHsPs5B4yM3MyKrMouZUeh2agp69dYzCBodmQ0RdNT9QexJTUHPZiL3IWhU5HZJybjMzaln6HbqDvFYZoqQ3C54GpK1mVHodkLQZMn6grwR2Zi4h26nvqC1uK8WyMFk43qRPtm+ZhS6nTJBl19//RSCxkxaCepqfxB7khT0f5/ui9btfP8/EDRa0kJQalE+dDspQb/af+Fnv/jii7//7JWdX0HQaMn6FXcdkr5mFLqdhKBf/XIzuC8//z4EjZRs3A+iJmWXNEO3syvoN3+pu7hovIKgkZDt+5UUZM/ddKHb2RF0+Y+mjP8m5kruqwVyENm547OX9LA/iD1JDPHquK8WyAEkcUtyD6m4XSl0OyWCzkULjyBojCR1z7yUVN5NF7qdtKDzW49Wq+vDAwgaIUk90yEhNW72DN1OUtDrw6LzvBCaQtDYSOqRI5LUuhc5dDshaHok8chRl9R9DC50O+kh/mLnBEN8rKTOFmD6D3KEbqekB93c6kp2ou6rBdKWpJ/YbJImzxmFbifdgyrivlogLcieZ97rpNljcKHb2REUC/Vxkv2bMmxJ06c0Q7ez24N+c1J38Rtc6oyCVOwasiYtNggJ3U5iiP+0drPIp+Q8yX21QBqSil1tCtLqEffQ7SQELW63++CLL97H7XbxkIo9l1bWOzCEbiclKG5YjpDs329pZb1BSOh2koKuFI98INyi2D2EzwYh1sEyU7xk7+BezowY1daQlAt6kUnvaHJfLZCaZP/c6Lw69WRSWwtSKujyd2K5afEyriSxJbNmKKI69WRQW0tSKuj1T8U56PI9CMqUzNR+bmdGwWtrTcoF/SF6UM6kfu85XpkhSKmgOAdlTao7z8a6UrTtxM0icZKKqVFn1TPWdkLQ+EjlvJ26ZhRhO6uvZYKu7wnF/aC8SNWJp+SSZnTt3H4t7UEX35FeSnJfLZBENJaVZFfco2qn7hA/x2PHrEidZSXZFfeY2olz0BhJo1XPkcpkQULQKEjLU89BZfIgSUG3D83tQVAOpO2p55AymZB0D1rtLLI8Jg11Xy2Q21itKw0skw8p6UGrjRvovRvcVwtklbqeJKn1nBH/dspICMqbbPSeBBnJ/iD2ZP8Qv5rfphZD3VcL5DqN0b1Daj/Iwb6dUpIWtLxT5Gg13zlpywlBPZKtk88WafCcEfN29pASQfvjvlogqblRgzR6DI5xOxUkBOVJkgtLW9J0Bwa27VSSEkGLHZZlm9tBUOckvfBZkfHtD2JP0oJ2d1i+nM1efQJBvZDSi5olafOIO8t2apGkoN0NbK9+82R19joE9UD2XHMXpN0ODAzbqUlqClpJCkHdknI5S9J2gxBu7dQndYf4PGUPeifPCnGTup7dv91NYIMQ6ygnSVf37j6svnb/ezNNsrf7jHp/EHtSIiiV5+9Uhrqv1iTJHjvLsZ1VbT2RBoKuHj+AoO5IpZ6sauuN7ApKfoLC5WvP0IM6JPvG9u3MiEttfZK6PejZbIZzUGdkj5/1a0ZMauuVNBniN3FfrSmRfUtLqewPYk/KBL3+ieRT5iDoyGSPn+1VTwa19U5KBf0RBPVBaveeI5YZFQlBQ5JGegavbRASggYkpX5K7qaLtZ1DSAgajjTpPMcqMzoSgoYh9WdG45UZJQlBg5A2esbYzuGkVFAsM7kjpStLirvpYmvnGKRM0N64r1bSpMRP9XNGkbVzFBKC+iYlvWfa+4PYkxDUM2mvZ1zthKAxktLB3WGZkZMQ1B85YHC3LjN6EoL6Im1nRkPKTICEoJ5I0s/J7A9iT0JQLyR58mn8DDH/do5PQlAfZMPONWnxiDv7djogu4KSzyRB0AFkq+8sSKsdGJi30wlJ96DrjRtkH5TkvlrpkN2hfWW8N51pmSmRpKD01jcQ1IIc49TTtMykSCtBEd3U/Fx/Z8rb1wyJxt5M27j/vUmE7HSfovdkW1uWJC1otUc9BB1Cdob3cnBnWlumpETQ/rivVgJkt/PcdV5mgiQEdUO2e8+J7w9iT0oEnecDPP0ZSRBUh2z52Zi486stZ5IW9PT2fw+PJB/UCUEVZOeyZmtdiVdtuZOkoNeHR2KlCeugNqRCT2a1ZU9C0JHJpp7UNSNOteVP0kP8XAzxWAe1IBt+Yn+Q4SQtaLkOig/yMiQzDT351DYOUiJof9xXK0qyNbh7KTN5khRUPrhDUCmp1XuOXOYESLoHnfdd6ISgBFnXU3E3HYPaRkTSguY5zTL6w+IhKEHW/MT+IKOSUkGFo1hm0iIzEz2D1zYysq8Hld4O6r5aEZGGekbbzkAkLWjf+A5B66SxnpG2MxhJCip/GgmCNsm6nbpPcsTYznAkKejyGILqkHU9iU+Ic1Lm1Ej0oNak8dg+QpnTI0lBV/JbQSFoGfNTz+FlTpKU9KDloccykyz2esbVzvAk3YN2c3VvNqs+jXvyglrMjAaXOVVSU1DxUdxXb+DjuEXqMyP9qdGwMqdLSgRtP5N0+Xr+v8cPIOhgPWNpJxeSFpR8Jkn0oqvVnTz9vW3KqekZuioTi/qRj5sP71dfuv+9YUk2lj3Z1zYRUlvQ529v/JymoPWZUT64M69tMiQpKPFM0tW9B9sX7qvFjmycemZ+ygR5Lp0ktZ9Javg5PUE7erKubVKkRNB2zmYiE53FkxN3trVNjNQUtBn31WJEStaVmNY2OZIWtJgl4Ybl862eu+1lT5a1TZCkBT3dE3vYzie/N1P31NN9mSC1lpnEKv3Ut77p0ZNhbRMlZYKKNaZpC9qrJ7vaJkuSgi6PDy52TsRAP1lBFXoyq23CJCnoarGf7a1Op7uBbX1mJLkjhFFtkyZpQRVxX62gZL3zlN6wxKa2iZMQtJVMS08utU2flAg62Y+h0dSTSW0nQNKCTvWDvLT1ZFHbSZCkoNP8rM7GzEh5s3zo2k6FhKBV6p2nxrMc0bYzMpIe4i+m9nHcBmP7aGWC1CG7glYPxWfTeS4+M+w9xygTpB5J96CKuK+WT9JGzxjbGSc5dUHNZkbjlAnSgCQFXexnvRuEuq+WJ7LZeRo95h5VOyMmKUFPCzcX+7J7RVIR1G5sH1YmSEOSELRaXUp8Fj9Iz4jaGTnZFXR5XF1AUmzCGHWaeoauDaJMfZmp6jjTXai3nRkNKROkFTlBQQeO7VZlgrQlu4JqDPHuq+WObI3tXsoEaU92BU16ktTuPXnXFmT/MlNyH8fdHdw51xZk8TUhaLlQL922IVZBGzOjzE+ZIIeSpKCquK/W6KRkZsS0tiC3X09CUOnEnWVtQda/noCgPetKDGsLcmKC9i57sqstyPNpCUrNjFyXCXJEMm1BldeMWNUWJEEmLKjOJU0+tQVJk8kKqnfFnUttQcrIRAXVvSGER21ByskkBVXMjJyUCdINmaCgjc4TG9hETiYnqNnNnqFrC1JFJiaomZ6hawtSTSYlqKme4Q8/SBWZkKDaM6MRywTpmkxGUIOZ0WhlgnRPJiKonZ7hDz9IFakv6NVvnjAV1GRdaawyQXoitQW9nL3KU9ABeoY//CBVpK6gj+9+zLIHHaRn+MMPUkUaD/F38ijf6yuNiTs2sEk5EZ6DNjtPuy1CQvcPIFVktIIOHNutygTpn4xU0KaeoQ8iSHdklIK2e8/QBxGkOzJCQZszIz9lggxF6gtai/tqSUOeeoY+iCDdkXEJKpkZhT6IIN2RMQkqnbiHPogg3ZHxCNqzrhT6IIJ0R8YiaHdm5L5MkAzIKARVLcqHPogg3ZERCKq+ZhT6IIJ0R7IXVOeSZuiDCNIdyVxQvSvuoQ8iSHcka0F7Z0aOygTJi2QsqP7tSqEPIkh3JFtBTe6mC30QQbojeQpqeLNn6IMI0h3JUVBDPcMfRJDuSH6C6s6MxiwTJFuSm6CNzhMb2IDkJaiVnuEPIkh3JCdBTU89xygTJHOSjaDGM6MRygTJn2QiqMXMaHCZIGMgrQQdO81Tz5A1QbgmYA9qOTMaVCbIWMjggg7XM/xBBOmODCvokJmRbZkgoyJDCtqaGYU+FCA5kuEE7YztoQ8FSI5kKEGJU8/QhwIkRzKMoOTMKPShAMmRDCGoZGYU+lCA5Eh6F1R+zSj0oQDJkfQsaN+qZ+hDAZIj6VXQ/kX50IcCJEfSo6Cqa0ahDwVIjqQvQTOVnuEPBUiOpB9Bte6mC30oQHIkfQiqeT9I6EMBkiPpXlDt25VCHwqQHEnHgmqceo7cIJBpkU4FNbubLvShAMmRdCio1sxIUi2zgEyXdCaowdhOVMssINMl3Qhqcuo5coNApkW6ENTs1JOslllApkvqCvr87dlrz7QEtdQz/KEAyZHUFPTmwwers9c1BDWdGUmqZRaQ6ZKagj7/7ZPaxx3Lf7TFqSdZLZAg11/rCXr11rPV83ce5l/dySN/31bP/p+HIGZRCHX5WiWoiNz9IY+4h/5dBcmR1BR024P2Cbq7e26tZ/hDAZIjqSmoxjloPjMar1ogQa6/1hP05sP7/bP40s7RqgUS5PprPUEV66AbPcM3CGRapK6gjbR+Xk3P8A0CmRY5XNCGnuEbBDItcqCguy09wzcIZFrkIEE7do5WLZAg11/bC0rpGb5BINMirQWl9QzfIJBpkZaCyvQM3yCQaZFWgsr1DN8gkGmRVoL2pedWJ2dBmemXCUFRJusyISjKZF0mBEWZrMvEHfAI60BQhHUgKMI6EBRhHQiKsM44gjbuu/eW7cNS3kq8N5s98Fzm5Wz2qu92ljt2+M3ZjGzoKII29x/xlUvv/3Di+darNx6q3zhixG+h/2Ob6+Jb0Md0gaMI2nz201Me3/3Yd5GXQhTJgXSZAEPFm+96bubNR/Qv/iiCNp+e9xb//26rVYh2+u9Bbz76xPcQn58lkqdPowja3H/EW0IIKp7D9pyre3d9H9qz+97PQcWpE9WLogc1yvO3vfu58t9r5/+e/idJIsTpU7znoGFm8SH+2byf94oJ9WwW4FfRlaDN/Ue8xbugIfwMdPrkvQcV7bz5q6Nlpqmsg5Y9i+d/ubxQ7+eggdZBqXbiShLCOhAUYR0IirAOBEVYB4IirANBEdaBoPzyrz93v5psIKgqy+PyA0xuPVK989svm68vKGTx0onix/yt9oaLX2nUMOlAUFWWxwd6b+yoZyfo/ED+aoKBoKp4FnTxcgNSd7iJB4KqshV0vnOyuj48WLz0fpbdfloO/sJB8efearGfZQeb710fZjvvywQtgfJNf2oKuToVP3lTRP7WPdcN5B0IqkqtBz3dE/8t9m89Et4U7sxvPxV/Xh8erdUrv5d7nH9PImgFU2+6PhRCVkWIMtXnvkkHgqpSTZIOhFyiv1vsHxWeXZQ95VE1Cos/q+8Vf84lghZ/mf+PetNi/6D4/1E1uotue8qBoKrUz0HnWSVOLuF8LW51qim+X31vXgzUm9G7knyvFLQcxU9abyp/SCnouogVBIWgqtQFPV0rVgpanCRu50KFoOvvUe6tqnd1Bb3+0SPxw3MXIWgrEFSVmqAXt/4pzjWL8fflRxc726G9+rP6XjWKd3+cEFS8qRrixf8uxLwq1zb/rz7EF35DUKQ/W0FFl5b7tJ0k5T1gLtt2znO0+Z6Y62hPkpYfLN97JFY8844Uk6RWIKgq1fnjzolYAcq1KdaA9tZ/I/q3atVInABU39NeZireVAh6VP6xWWbaK0vHMhNilnHXzosT0m0PWv70bRHXP5z2CA9BzTOaoOJctOwhhaDlOWh5cXNbBC51hq5AfBmvB51n1UD+XjWLz/PV0baIxc9HKiraQFB+we12tUBQhHUgKMI6EBRhHQiKsA4ERVgHgiKsA0ER1oGgCOtAUIR1ICjCOv8HlnPmbbDSJaUAAAAASUVORK5CYII=" /><!-- --></p>
<p>A manhattan plot is made by setting <code>plotType = &quot;manhattan&quot;</code>. Significant SNPs are marked in red.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="co">## Plot a manhattan plot of GWAS Drops.</span></a>
<a class="sourceLine" id="cb14-2" title="2"><span class="kw">plot</span>(GWASDrops, <span class="dt">plotType =</span> <span class="st">&quot;manhattan&quot;</span>)</a></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqAAAAGACAMAAABFpiBcAAABCFBMVEUAAAAAADoAAGYAOjoAOpAAZrYzMzM6AAA6ADo6AGY6OgA6OpA6ZrY6kNtNTU1NTW5NTY5NbqtNjshmAABmADpmAGZmOmZmOpBmZmZmkJBmkLZmkNtmtv9uTU1uTW5uTY5ubo5ubqtuq8huq+R/f3+OTU2OTW6OTY6Obk2ObquOyP+QOgCQOjqQOmaQkDqQkGaQkLaQtpCQtv+Q2/+rbk2rbm6rbo6rjk2ryKur5OSr5P+2ZgC2Zjq2kGa2kJC2tra2ttu2///Ijk3I///bkDrbkGbbtrbb29vb2//b///kq27k///r6+v/tmb/tpD/yI7/25D/29v/5Kv//7b//8j//9v//+T///8AQlJEAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nO2dC7sct1nHJ6GxW2hPr0CBpDRkk7YuUCAppWkhJHUpdca1ObG93/+bsLcZvVfplUaa0e55/88Tn90Zjd6LfqPbzm6GvcvVsYatHXC5YnJAXV3LAXV1LQfU1bUcUFfXckBdXSsL0NHlWkkOqKtrOaCuruWAurqWA+rqWg6oq2s5oK6u5YC6upYD6upaDqirazmgrq7lgLq6lgPq6loOqKtrOaCuruWAurqWA+rqWg6oq2s5oK6u5YC6upYV0Def3H37UwfUtbasgH7xZP/iu186oK6VZQT01U9+50O8awMZAb1//1eXIf7RQYne1uWqrhSg7z05QOpDvGttWQE9wPnqo2mVtLXTrocjI6CvfuaAuraQEdDjKt6HeNf6sgL66sO778wL+a2ddj0cWQFF2tpp18NREaDDQceL/a//bf3Xe1BX13JAXV3LAXV1LQfU1bUcUFfXckBdXcsBdXUtB9TVtRxQV9dyQF1dywF1dS0H1NW1HFBX13JAXV3LAXV1LQfU1bUcUFfXckBdXcsBdXUtB9TVtRxQV9dyQF1dywF1dS0H1NW1HFBX13JAXV3LAXV1LQfU1bUcUFfXckBdXcsBdXWtTQC9/Pyjy5XUFoAOgxPqMsoBdXUtB9TVtbYA1OegLrM2AdTlssoBdXUtB9TVtRxQV9dyQF1dywF1dS0H1NW1HFBX1yoC1OVaW96DurqUA+rqWg6oq2s5oK6u5YC6upYD6upaDqiraz1cQP2p6avQgwXUv3dyHXJAXV3LAXV1rQcLqM9Br0MPF1DXVcgBdXUtB9TVtRzQpHa73dYuPGA5oCntdk7ohnJAU3JAN5UDmpIDuqkc0JQc0E3lgKbkgG4qBzQp53NLOaCuruWAurqWA+rqWg6oq2s5oK6u5YC6upYD6upaDqirazmgrq7lgLaRf/xUSQ5oE/kH+LXkgDaRA1pLDmgTOaC15IC2kQNaSQ8X0LY/3OBdaCU9WEAb//SNA1pJDmgb3TSga4bmgDbSbfO5XnAPFlD/8bBiOaCuruWAuvqWz0FdTH0tutZzxgFtpMpNuOtqW2BFZxzQNqrdhA6oA1pVDmgl2QF988kTB9Ss6k3YE599zkGf3TmgGeqKpyuWGdD7H//UAXWtLiugb3756/MQ/+igZG/rOnSgW7twW0oB+uyDBzsHLRmsu1rSXLWMgN6//+VDBbSINQe0loyAPrs76gMHtOVFLkHWIf7hbjOVseZ8VtLDBdT8uJ2ztqXsgAJt7XQN+f/t+DrkgLq6lgPq6loaoK//8Ic/3jSg/pWP65AI6P/99vGxf3nrm/97w4C6rkISoL9//Gc//JvPPvuvH/75W3/ngD4o9bdjIQD6+7+dB/fX//lNB/QBqcPPFzigf/o3yOJL9M4BvW0ZAd30S3Ov/xvD+D/CWmk171zrygaof+3YtZWsHejmgD49ruLfdUBdgjoA9Onbn+/3X33/Ww6oS9DmP9zw1fdPnefzI6YOqGtL9QToLX2206yT6W4fqLHkIf75Wx+vP8Tf0qfjzaZpHe5UtpXSgw6TxE60jSsO6JYV2+2va1DuQRNq44oDumXFvZpngG64UX87fN7sHLQDQPd/+hiy+Cf/qNMV1AOg+9+Ch0V+K66TVvXQ1ZO6mIMeH7f7l88++0Wzx+1uaSx3VZEKvgRo6weWb2k15KoifeogArpv+5UPEdDVmfWbpCPlAxrVQm8kQFfvVa+tG+9pf76+LyWAPh/UJ5qWuiN3oA5oRFtv0EO18CUf0Nf/eNxuevmNBp8keQ+ar4cAqFinCuhXf32cg77+p5UA9TloXA4oA/QHq/agq2trJ3Kt98NnE0LzAV04B43mvwM+VUBX8m3r+2OZ2hAqHtcBjWi6WM9x//lXPOSH23Rd/ScophUnHEsAjSR5s/zbE6d2oPh4o7bITVBPI3wfgF6eCY09D9ohoHPmSs2vBWimgx2skZAD63mj96Av/0L9KOlybYzC7TrQU+rKbxBhhN8ajS68aO5B9iJp/zT5teP+plHLAZWqrFPRQiduHFCt/gigupY60xDsMMJve/PUbs2t+XxQgK5BT44F8XOtZea37/Gq6xxQy28K2AENX5p751oBzRB3Z7mDNwjoSQ3jypqDTr8s8vrnIqELXblaQO2N44DWktKDTj/cIP92w1KjXfFpBzSndVq0YwfM3wagffGXlnUOmtU6VRvy5E8XvfLqLsSH+P3Tr0mboccLV/yYs4N2OSmzB63n9TmjXQB6kcWTOhTIgJ6fFHl3//StjymcF0BX/BSpn4bJ60CrArq7qFaVy2RxpRIGCqBxxc1vAmjDtiuqujKgffG5ENCsOBoAWnsOasmGWKaOH4Vg1J2D9kTnuBDQvFgUQE+/sKz9uF1iDlpdtg6UFTL35PFiZWxU5qkvQIk3Sv6mw8uebpABvbpfWF4CaKJcERvVgeqbz8TT6UueDxMBTf2AbYc6xEwPnTJjuDRVTqja4k7JZclKK9dYJhJbIn/8dFEcPfzCclp5/Yh9hK8+XWmxpOlmmF/Ug+bmWgS01yG+URO1mE7XdnXJMr7aYlF4eXyXqJ/ymeeNDKhhkbRUJVnrpg8xqA2fJXXWGSGgdeRHnlu1AI0rx0AtR4+6JkDrys7ngt2MtAfsJXuX0tUCOpg833SM21JmCoSCbQDdCScsWj4HTf0fFJoAOlmsUHPSVoa66bCXABplwpwZBqgyI62tjXpQuQMtm5ZmG8upvBNC8wC1O52RGtppbv2146aAsrxAQHNiN2XqIQGaSWhhah4eoCPis/ZDQwtG+E4AbfUgVem9u15eNEC/+qvIp0iLra4MaLFKx7JqY2DzvQ5Qf59LSRXQv2wJ6CkZwiz0+CKvbVvfykWk2TeFEiqbmJdZhqbiNWSHtmR/ZiNAR3GdlPB1C10joAtsnV/Gnc+OzRCEXmU/gCJvq1hYqqMfnQHaMDUlgNrcuRlAi8Z4tfKFNUx+dDUHbTrlRiO8CVCjO1cKKINoiqNCK1hGxgVJW1vGrq2eJCvhWC6gVzoHZVoXUKUIOCj4sQ2xJ1dPpje8Z4DpbEDtJvhBFdDINlNoxJozeLzRlBGCVldJEXRU4tPaIDVb7sLnhdBq1eYJxg6moBX9kWvTAI0oNGLVNWaoTI3bnhDTCD+VwXuBxzfK5Xb7VZtuBtRuvJZpVCmv9nyoEgQPClCbOWp4eq1tCW4EqO178XwEriu10ioUqPH1A6i0FCAeV0z9IAM6Th2ovCzJGeFrT88SVU5nc7raOp1fDQr08AoA3bfhE7sb3JZORK4xSgOUHikFbStA0z0tlNyC5akslu5zCaCW36iPyXxVRqpzgSDjOj9bWG+pP6JxUl2UPQxorC5sihfIdr3WCC+eWQFQpX8yXNsQ0LT55JZCXX+gXaVXSxLK7KZaaLiIVlSz87fhq5pcAKidT+HZT9vV9kzVyiob+AMUWQaAP0tGzHx40PlkkgGhi2c1EQsLru8b0AxV43NeJSFAsz+Vh3xK16nRk8kHLpbphLUHZQtDu4mlLqS0AaBspdyRwM2zENBZ8nWR8HEHyrtQu2lzB9qwPTYEND67QN1AvIBJyZapct9jQKd6G/Sgc/2pJC54mCltA4/wbTqMhdUuATRmvsEIHm+cucAyUOEIT2quOgc1csH4LOlALb110wFtSdULAVXjqhmwqf8CGB3/LrOPCS2R4C+qsKjjMgAqreF1G7CPnSY0VmfsWpTLDQFlBZQrdrNitWFAFxI2yNdnNB/3F1fYCFBUIh/QmIFydm0xKvW3AtS0z4i3Q7W65imgKUAzoMJp3F4DNZkzwsKyu+mRCtKFam5Ea7Wbtc5BwYudnut48NEoWEsnPQdaCOiC6cXZbdBs4CXytWR1kgZUOA8uuvCJDYf3aYfwHFSYdJQmLm5aBlQXTP44zo+kCEairZCwNPEZK9MK0HJFACXOlgwuJj51QMfQWsgP2EtHxQGV56C5SplGZ1MmyPmhFaCWMtUBtec3OglASIAOdOlk3dA2LGHk2NxSEDXFP+4v6m55NK0AzRH1YYCESpbNNZWUUeovBTQ9vzE6xk6qGYqKDMbJjIlzgHBowDOy1GRUcHkjQHMSJwB6nniLlcCDrIAlGAyM1dGlgKYTndUUkTtYvWC+Dhw0uqUVOZ8LHSh1RySW+rAQUPl8ovPOurWZDzGXYM3USs69djFqc3QXfs2+D0BL+IS7n/NR45gDPKQtdckiqzvht9CGZYAqBSI9NXAk7i20IbyUFQE0qwvKAfRUqBBQ0GGL2zXqu7RHOTnWADURGvwTBztYnd1x4ph6va2HH1hhsfPGs+OMThRZi5agd5wZUCGzqwAKLWZHG/VoOaDGLvSiFKApd3ZglqGOgtkahuBXLBiSsWaAjuqQEM81OZk7wtcBFH1AIrY4uUI9sVMmfbEY5lB0t6JeRAHF94vkl3J6IZ+I0GHgnxogQ4tvDWMPuqMvx2m4stU8Z9ro5K58DhpxIKRW95ifDQeZ88abjdUnlxzC3Fmdg85XC92TYDm16C0UBDTeM87ndtoaPGplDD3LVIloQRzjE6zB3EbaWNWGgPJlCb65pjQsHi25eTzFwwGgv6wDtQKabjarr6eE4huFi6bKnjF6o8aixFDG7lpsgMaT4+BiQFGuqwG6A51Xa0B5E4VDQruwGrUOdOHGdig3ZAOakTIFUHo1Poi8ybwZIpNp6dhyQKkLljnofB4nJ6SnOaAgWTFAecMYzXCn+Y1sdXZdQBVDQgcKulCbrdjAwY1ebNQAVO0NI47SXVSUHQaoIXhrmhiWmFRA7zxZ2hVMKQU+6VTI5CtYI7FU8Fsn/CneDE0DiptlfmULSRMzOtmoAChNN+wiNYGuFlwEQ8/nUx1yBFeifHJCc1pcFk6IoTGnDF18EdyIuFXsrx7rLtKBHl8qLQ4bOGlY8mQRoKwfnA9bulBUirTCLpcMfEdzY5ID+BVDCGx62RtcK2ftNBUHpWzE/IInMgybIg2Fpnaa7yM9DPgi5TC0MjGXAvT+vbu7JxRQEbNRARS9l/nEzZDFRRGgLHOyN6TLsLnBlMdnRUBzbg1TlLiFxsunInKLs+xq3rBYLm+NgL766NP9/Y8+1QGFuRDdxXcQC2mYmYiyFk8bDEwyzQ7OqZsTSpyhIrZEL7RzcsNo8AzIQbEfbwVoKuWkCwHNLkWAARU4VoLJA/TF9w7/fDF1odQF7FMcUOQo3SILHRZy0qwpbeSwmBJ2b5NTcMVG6Ii4FS4IDhF7yCedHty+1CQChLmALOcAmtGFjoFU2QLBFjS/PsaT+2/M2gc99qL7/aODjr9vt7/8DuNwfn2p5fSOH4eFB3LxAM6HfuJ4CL4+C73BOp4ipeOabau+7oM32JWYnRDCDhdEVnAaFNdghngizl4pDoiBGkQTLhUIFuY4RQOx0HTjmuF0BG8++QAaGcQ7BN4mao9PFe6v0LjwjeX23vHScSHTg9Dbn4/ukNJ+4M0HWBBaCYHzPIWzMEHEZCRcdlQyoSmVQWL3/HKqXsgir0HxRY3HDOirDzGfdNkW3mHyqMsyoKAd6AASA5Qwkw+oPOoA3zCg6Tko3pea/5X2QdVF7XzWDig4ybKgATofDaeTGZQyodxrOfdFpPmsgN6/N6/hY4Ay8qb4Q8I5pzia2U3uMo0At9AFBSOk1DaLhdNppB+WDpQqgErNyLMnz0HDMZoI1oVqCUAvkssumGZekQ1Q2RklwTvzNhPgEwNKc84BhW9BJGlAode70KfSsMiiQIpTSgqxjdOp8ZmxIxsKC4COUuDEN/Re9oDcniF8vo3B3cThj+hqeZ8eNYRSEdkKUa1KtUt8Wjfqn90d9UQGNPwr8Cm0P7wcYQx8RekCmKLGF5pNaR8hLZzPgZwVALVuhbI0SzM0HLeuIfp94NkEWlhLkbJawXCGK1KntvBfMRTc8GIsWheqhGUc4oEohSKMKUCZr4TQkfFJDs1hyYCyqllepiN2QC17MLN3oGT0IgOhaUDp1qQFUG3fxw5oKHaqguSRNa813NleKaCI0FEElB5Et5NQZK5lCD/oQfFgwz7KHE7vDvohA4rdH0buTAmglqL8TlBbbPYwBSi5rbGhlAmBT7k7U/rRIaQQVIcqRocV93k47QCN/oAyQYNeijINaCWOa7GdUzdIinmBHIZ8wimlDgi9f6Sigziox+gJLuIbRbTPziShCKGG0zE+0aovhIjTixsU9dOaD6OQ2WqA4gBFQAd0I4GcsMggFyP8w3pIPpuGfwmg4iSUACqEQVFT6eA7D/gIajBhBZSgB7eX6IN4PAVFsEABjYY4IkB3Qh0wsEgLyLXjRNYCVAh4IHdSeEm5wYCGXAM+0XyO5xBcMV1EvBPSIwIKDwHUWA6l5SbML040t6e2FMtkmJjrKxh4NNCTA+hczg7o7I8KH0lszB14P2NCZ+iqAMqAHTmgwtVzBCPbT5HfkSRiQCGhaK0qZA97BSEa1SdXIneImnmYmGhbTSdocndQohG6sRH/siXKQghauftIiMgZ9UYggMC/SpZGIcsNAJX85C7PPODl304RSwzInQToXCEmb24HvHolSQSnpiqBXRugyDWUKJY1KVkkS1E+hd5NLqcZDEELV8LkwkOFgA7CN6iBwxjQcTGguJ0zAcV1RAEVEgMnXphQMMTzGSbJF3YS/5GdiQO6Iz3B+S8MlQ7e0AcBUASOhp3CjlSUCSZEuvByQG2HHfSZVYwzP84LPp6+kDCUwGJA5cGT88mWxwxT3KWAbDAgCJogZ7AI7EAZn8htvphGMcloiuO55AdwKBKv0J7w7cgATXAHuYkWhLOJyeAiQFlISpgURaXbv7wuAXQ/OwTjnF3ieZCH0uA+D4QDCicmEUDP72jNSl8JjpPDp9cZgJLWhAGAtqLrQyFP89G5AAdUuzvoBBTWxV2lHmiAwjloHqC8G5hGeDS5J9URNxcAGkyqQdMCCqAoEOI26jchIigeElqATCR0JOQOYnnhSaYdufWJdqJYgzGXpNyFtylA50McGxAqc1YCNPa/pQeHBUDFhWgYMkcajBHQXfk2kxinnAqSBELoOEJCWLql8X5k0YSiI6qP0wAzRZwZKUHA1CitZoUmVAilVpXGpKmaXywAVDbGemxqRwxODBRaktLCWppGIgUHDtQAlDW/4CQqjEEAMcxe7sgrMr6SNOKTI2seNsmAr8OxgWiulNwnkS0lUcyqAo2QsvQcNEStYSPbmhMPHFOtgCzQOFPxwNM0GClx9EghoCTfJKuSj+LISwjljRwFlDTSfBJmDKdPAJSeRXxqkiChRUgbMjCkVPGjgifQLNtkgyal5SAyBc7J7OAA5XsQt6NsIwooyeC4FFBE6MDeyLmQuGSuy88Ihz9jmAqR0zg8WDOywloEzJ2QQ4oz0eabj7M8Qw+ApbkNUZ7oRJR7gsziLdcdNkkN0DZh6YhHSIcRFBwJEtuQGxlWTzebZhUAypt5MaBKU6iNEhlPdxBQ6Skw0jrEA1AiZkNoPnHGgduQowFTRltYSwpqUuZEsDmGLkxrE2Z8B+fyNEQ2GKPgqPtsMUyi4flD4U3KBxS1JHIKtTrJxsgAlYhNAjG9U4pd/hWNBFe4E+xuj/hCmo+nljkX6mVtKaQvBSg2qjXztDijlWJASRvtOJk470K6aRMSE7wleA6F3Ib32YCiXI/hDZ9ZQjglhzmzSR5wbtTikhXJhzJAUTsSy+IF1ChtM5QfoYF5MmRniE0boOGP8CEP3QbliWDRpQCFYQgpVJbxuYDOEWq5n10lLcBnYEBi6+7kXmIEw440vpFsyQbheoUlU/MG+zS/BU1Gu1FiTMwTdJI1MDMsOyNlgNqggIZ4hf/7Ix3a6f0oWSO5hu0djDATJBR4pLAHlVPOt1Gwf1q3z9uB+Bz+hINK2+zo3hz0hzUcbD98XPEGJVIGlOw9CLteNA0EHeopwgHFqyWBRMj4JAMHXRUqgNIczC9YRmEkOKSp/p0Eo3RoEaDyrGrEToolQ9OQdhASIOYJ9aByFypMKpgD6jmpIVgm2ZJd3C9ByRkEUmFriiVwNpRuB06AYc2yFCtSeNqMfzouZFQzmfioEyX1rH0uoOIT9dN/wBEZCH4pKKRlgwbBC5D3aVPh2HyOnIlUz1IpMCP5M+MneAZTGR9atG5H7EF1QKV1g1KlGFswSFxHdpELs5HRDOgMXQagke98QDzlOZ90qdwOyG38VitHW0fwEeRwOojSKzWW2EC4AXVApZVuDFCxBDeMD+zY5oz+oDttJs0O2tNV4ifpHudQsBHpHa0EZTXMHRoACo5IORAuDS/lNmZdqFrqErZmZb6lkeOKXzRvQvuMYAhUCd0JKxYaOvRC8kcwrSUH20zxGV0DxO3MJ1mm+S0AD4DMIVRhWoOREkClXHOnUjkHBdUMgWyob4SMST0o3U0QrJNRNuUJyHPCMe4PnYtCp5KAklaU7wxQ8TJAyY4WOSsEJ/DJ2iAEEQAVB6NKX/mgnThJeRRQFIjSwFaJk3aaHNQ4sV5eNgGSh5sPOkCUDB52MBZPMKCSf9CGDqi6ig9WYvOp02keUxjp5ZhRZ0k/eyE2awFKNmowrREIyJ0WSUaOBDvQL+Sl7lvKSqr5Yv5gvxAzJk+gaXGfhuRYJhSuF7gdwE9kUqUtO/XM4urpOozeFdUAZWsQqfG1TxciTVEiLTW8WSKAJq3UAzS7B43OeEcy8iqA4uClLjQdoDDEz3FoqWUmBKPgYGVAR3JAKqRcrzRFiUDKiXV8RvAmvaWAN8ft/jAD0CMAqOC34ELKMk16ClBsSFljGy1puZWCUQAN7/MBFa0ORkAjPb8wyOQqPQdlLmve8Dud2gq3eqoV0XJWdIA0bHJcSZNDrFjoTK3iDaZCzYqF2Yi6xKNGawCqzIpFH3U+R6klCqUZYQ7GTpO0hdchkTZKtfBxa4aXUU+CcXMGrHzCffrdzjIHTVhiCQ9WKKBgSt0CUDFe1CsI6YhmaKFS9lQWZG9QzqRFKC2DhM2KfsDtD0NesOW4TcanHjo1kL4PdrQzSN0DkxWwlIdZFRKZDag6xrPkJwmIZGihqvgWquPtJLRfGlDFJHJFLCOMg9wMvFci/dr5UDrgMkCRNXSYWdEMSM+K1ATUuEiOJ2ihLMYy+IyN8HpmmTtaOpKACneHaCfs1ZDnN5K9GgtZAjT2kRW3Rg9xIxqg+G1tQOFt2grQ7Lu6SNAgyGMBoOIvnQBpH3DIeSnp17IBNT3OJOf6PIlg9rkRpQtdDGhyjG8DKJykRAtKSSuRYh5+4pF8zMrmD+lC1d2f2U7xwDu9S0ds5VN/0kATNTHiz4uZ6gOKdveShZUEoWTt0LCWbB7dMfSnyJtRBVTnJmlKezxA9qQcUJqIiB07oGyZlAiWWEgOCPmAJtt3BJsnqbJqhpRkWaT7tdiZ0GhSD1oMaNxLzYeMFFBAxUxQI7acM0OpYLEF2xhfuwe1TXkSObJIDE71q8wV8YkQKblKe6bt5wFqmOMouz8RG9ROKaCkauGjmpQNeiQbUC3LzJPknCeaIZvkNaDVRaMzUjNZBqeLQEYUU3C7WLirue0Mo1PVE6Gj7ogUoRK+ainZ2KKJWHC5gMbNk6zYfJYTZBfLoMm7Zd5kTDuSJi+J0kcdZttuc7aAZYnYNEPUvr6gZRVXZsnhGoAalUy7oiJAF3pTAqgm4bHQmCsWs9SAhc+i/Szl9tNbPjuWFQBtN8JfolSTlsyX9WkRbtMOqOFzreBMcnK4K+lBU5+Si3Zs8dFUxj6qEm2kDOQCaie0/RLpEqSas0VKGK0JaKYrBYBKn3emzRQAOk2kjbFwE4vnoP0BquYs236GLzWH+FxXRvWNbhPBaZ+D5kY3W7AFI5lYD9Cc6YCYoVLlmi3wZc7GGt5Q2yU5wM0nN40QY7alJCDUQsrEPhdQc1ojcx2D70uUZ7XMGdsS1+hNxs7MLmzE5hglo3t6H/QSYzK6xD5o3IicwIVPM8XN04wYS8sJKlWW0UJvagKaylOFFJCP9iybBYWAxsIkViwZnKhrBKjZeTFBxcozW+iNeRaaMkTzlNyoL81AaBG5baQIC6Mz7I4Y7/BsQDMINZZUE1SqXLuF3lhXSUlLOaPikhQAQC12ltx+pkXS1oAaC6oJKla24VxvwMfx9GCRO6sDarPTGlD7B0mdANrVKn7FHjSVqwopMD1eJgRYHJ3ps4AtAS0Y4a+nBz2rJqB5PVuBTbyKN9qpMcNO2mA1rvOlufwl0jUCWsmd+pNQwUSfgIo/7YjeNQJ0ww+SKgFqWMbX8sb8LYnyHIwGQnmAKwCaVC6gmQ61db5Kzkq9uUZAc+zU2uZdGIoDWuiM9Tsoa/iStlrC584y3W0eiwNa4Ix1+lnJm+UpKOhAW4WXa6AZoFc+B42bsPPZA6CWFVKRmYL4sutvCOgtPyyypjeLjRoBLYq5dTD7XEBLvjTXyPdaOWvqTIPnQQtsdgNodv3ZgJpdufmN+tXcqWHTAa2gkvxUyFljb5Z/x7SGzZJtpmrRLYnlFgHta4w3GWsP6FDw3GkPG/W3COiteVPNaPZDKR1s1DugZd7YN0Jbe2IyavoJjaIom8fSEaBX9rhdRW+qP27HvilUFnIHQ3zDb3VmK9/5Kjlb4M0qT9vVG+KrxLzMVH79DwnQ5T+/CFXvYZGlntQyWmBoKaDpHD4kQK/Um9WMrhBffv0OaGNv+tiobxdyaxO5gJaF3sj5Ginr35uKRmvvt/rjdmvkrHdvKhp1QJeoKEEVcta7NxWNOqBLVJSgCjnr3Zu1jK4RXr4BB7StM7cOqC+SWqessTO3DmhzGw5omTPX9Vl8KzsO6Ao5K3PmBgHt8isfvlHf2JkrAnSN8PINOKDde7OW0TXCyzfggHbvzVpG1wgv30AuoEWht3K+Ss6692Yto6uEl23AAe3em7WMrhJetgEHtHtv1jK6SmueKgUAAASrSURBVHjZBnIB9Tno6t6sZXSN8PINOKDde7OW0TXCyzfggHbvzVpG1wgv34AD2r03axldI7x8A7mAFoXeyvkqOevem7WMrhFevgEHtHtv1jK6Rnj5BhzQ7r1Zy+ga4eUbcEC792Yto2uEl2/AAW3rjAO60IgJ0Fcf3n33Swd0I2/WMrpKeNkGLIC++eTJ/tn3HNCNvFnL6CrhZRuwAPrqJ7/b3//9764e0Lq/zbTUm5aerGWnE0Dv3/9y/+qjTw+vHh2U5U9OaZvDj8pzBu6utF/NvTE7U2C+yFpRnLlmsg1YAH3x3QnQ7C60HqD7VCTpnGUBaiN0DUAjnjigpAfNInSP3UxfGEnQ0XAiEpy0Pbc4+X78Y7lx4kbO/jzaTX7R08ibuNLOXFLPrORwM2VDtyZFmMy65rFmRqkGGQCh7i2Awjloph6VXLRCjbX8qlJPeSUlVxZZy78o84p48eQq/oN5Fe9yra+MfVCXa33lPG7ncq0uB9TVtRxQV9dqBmjh2l+v7727uyeLa3lxd/edOn4dPwReqGd3pd68+eTu25+mi1Fj+Rk8ZD3Tw7wrTpRE1zmtAH1RC4SLjpux9z/KbBSmYz4qbUo8W367fFFcw+HKFwVr1+xrjll/lnXRqw+fZFxxogQ+78HVCNAvvv0fdXvQF8cIylsUqE7Xfv/jny515s0vS++34+50yWUf5Ro8fU6TZQx/spPQmZL4XvvVDPH7kgRLqtKDvvnlrxcP8YeRrXDWcv/+r7KH+H1J5Pk9aBagZ0ril1wRoMcPDRbr/r2CluV69sHyOehxxlLWi96/9+TUrHkqub+zt8FPQ7w9xUdK0PMeTNcD6KsPK/C5r9MPH+iosEg6qmjWktlNXVQwaz3eQy/yFhOHRdI/2O+6G+pBj71GHVWYyZ6XxDVumCJnXv2sBNAv8v2Nd26aMmat9zczB63DZ1nCRS3vQY/OvPn3oix9UTDEl8wm8nvQ06zVPtc9UhJ/3uNaAC3bxpOqqTIHrbUPWujMYWqYvYtXdGO+yHbxcEXGVGK7fVCXq4ocUFfXckBdXcsBdXUtB9TVtRxQV9dyQAv0dBiGtz7e719+/eOtXbl5OaDZev3zr/3xCOm7DugKckCz9ZsjnwdCv/ZHB7S9HNBcffX9d6eXL7/+i8Ngf+hJv/HPw9ufv/75MLxzPPivj4fhWy8fH8/sLwf3x7fgPS/09uehkCvIAc0V6DZfPj50pk/f/vzl43eOlL1z+u98cDifmQ8eL3r5+F210NwhHwptGV13ckBz9fIbn88vjzAdsDr9fX7sA58faX03nJkPXi7SCx06ZlCza5IDmivYg379vJQ//X1+nJpOb6Z/poP735xHerHQ0/NPF31rKuQCckBzNc1Bv/rBxxmAHq87zDNlQM+rrv1UaKvIupQDmq0LT8e5JwL0uDP6PBw8A3o5eLrwgLZeaBJYg7n2DmiByD7oBCheD13+mQ6eGAXvSaFDhQdKp0JbB9iVHNAC/QZ+kjQBGnaQwJl5m+n55RK10KlrvRRyBTmgrq7lgLq6lgPq6loOqKtrOaCuruWAurqWA+rqWg6oq2s5oK6u5YC6utb/A6wR3ZSKn0XzAAAAAElFTkSuQmCC" /><!-- --></p>
</div>
<div id="kinship-matrices" class="section level2">
<h2><span class="header-section-number">3.5</span> Kinship matrices</h2>
<p>In the default function call of the <em>runSingleTraitGwas</em> function a single kinship matrix is computed using the algorithm proposed by <span class="citation">Astle and Balding (2009)</span>. Several different methods for computing the kinship matrix are available in the package. Kinship matrices can be computed directly using the <em>kinship</em> function or as part of the algorithm in <em>runSingleTraitGwas</em>. For both the available methods are the algorithms proposed by <span class="citation">Astle and Balding (2009)</span> (<code>kinshipMethod = &quot;astle&quot;</code>) and <span class="citation">VanRaden (2008)</span> (<code>kinshipMethod = &quot;vanRaden&quot;</code>) or Identity by State (IBS) (<code>kinshipMethod = &quot;IBS&quot;</code>). Users can also provide their own kinship matrices as input for the <em>runSingleTraitGwas</em> function by specifying the parameter kin.<br />
In most cases using chromosome specific kinship matrices considerably improves the predicting power of the algorithm. Chromosome specific kinship matrices are calculated per chromosome using one of the methods described above, ignoring the information on the chromosome itself. E.g. the kinship matrix for chromosome one is computed from the markers that are not on chromosome one. Calling <em>runSingleTraitGwas</em> with <code>GLSMethod = &quot;multi&quot;</code> assures computation of chromosome specific kinship matrices.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="co">## Run single trait GWAS for trial &#39;Bol12W&#39; and trait &#39;grain.yield&#39;</span></a>
<a class="sourceLine" id="cb15-2" title="2"><span class="co">## Use chromosome specific kinship matrices computed using method of van Raden.</span></a>
<a class="sourceLine" id="cb15-3" title="3">GWASDropsChrSpec &lt;-<span class="st"> </span><span class="kw">runSingleTraitGwas</span>(<span class="dt">gData =</span> gDataDropsDedup, </a>
<a class="sourceLine" id="cb15-4" title="4">                                       <span class="dt">traits =</span> <span class="st">&quot;grain.yield&quot;</span>,</a>
<a class="sourceLine" id="cb15-5" title="5">                                       <span class="dt">trials =</span> <span class="st">&quot;Bol12W&quot;</span>,</a>
<a class="sourceLine" id="cb15-6" title="6">                                       <span class="dt">GLSMethod =</span> <span class="st">&quot;multi&quot;</span>,</a>
<a class="sourceLine" id="cb15-7" title="7">                                       <span class="dt">kinshipMethod =</span> <span class="st">&quot;vanRaden&quot;</span>)</a></code></pre></div>
<p>The analysis can be performed for more than one trait in a single function call. Not specifying traits at all will perform the analysis for all traits that are present in the phenotypic data in the gData object. The results will be combined in a single output object with the traits specified in the column <em>trait</em> in <em>GWAResult</em> and <em>signSnp</em> within this object. Running for multiple trials is also possible and will generate a list of outputs within <em>GWAResult</em> and <em>signSnp</em> in the output object. Not specifying the trials will run the GWAS analysis for all trials in the phenotypic data in the gData object.</p>
<hr />
</div>
<div id="references" class="section level2 unnumbered">
<h2>References</h2>
<div id="refs" class="references">
<div id="ref-Astle2009">
<p>Astle, William, and David J. Balding. 2009. “Population Structure and Cryptic Relatedness in Genetic Association Studies.” <em>Statistical Science</em> 24 (4): 451–71. <a href="https://doi.org/10.1214/09-sts307">https://doi.org/10.1214/09-sts307</a>.</p>
</div>
<div id="ref-Browning2016">
<p>Browning, Brian L., and Sharon R. Browning. 2016. “Genotype Imputation with Millions of Reference Samples.” <em>The American Journal of Human Genetics</em> 98 (1): 116–26. <a href="https://doi.org/10.1016/j.ajhg.2015.11.020">https://doi.org/10.1016/j.ajhg.2015.11.020</a>.</p>
</div>
<div id="ref-Segura2012">
<p>Segura, Vincent, Bjarni J Vilhjálmsson, Alexander Platt, Arthur Korte, Ümit Seren, Quan Long, and Magnus Nordborg. 2012. “An Efficient Multi-Locus Mixed-Model Approach for Genome-Wide Association Studies in Structured Populations.” <em>Nature Genetics</em> 44 (7): 825–30. <a href="https://doi.org/10.1038/ng.2314">https://doi.org/10.1038/ng.2314</a>.</p>
</div>
<div id="ref-VanRaden2008">
<p>VanRaden, P. M. 2008. “Efficient Methods to Compute Genomic Predictions.” <em>Journal of Dairy Science</em> 91 (11): 4414–23. <a href="https://doi.org/10.3168/jds.2007-0980">https://doi.org/10.3168/jds.2007-0980</a>.</p>
</div>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
