<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An R package for fast GWAS â€¢ statgenGWAS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="An R package for fast GWAS">
<meta property="og:description" content="Fast single trait Genome Wide Association Studies (GWAS) following 
    the method described in Kang et al. (2010), &lt;doi:10.1038/ng.548&gt;.        
    One of a series of statistical genetic packages for streamlining the 
    analysis of typical plant breeding experiments developed by Biometris.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">statgenGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="articles/index.html">Vignettes</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li>
  <a href="reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    functions
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-seedling"></span>
     
    other packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://biometris.github.io/statgenSTA">statgenSTA</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/Biometris/statgenGWAS">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="statgengwas" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#statgengwas" class="anchor"></a>statgenGWAS</h1></div>

<p><strong>statgenGWAS</strong> is an R package for fast single trait Genome Wide Association Studies (GWAS)</p>
<p>statgenGWAS has extensive options for summarizing and visualizing results. The example below shows a selection of what is possible. For a full overview of all options it is best to read the <a href="https://biometris.github.io/statgenGWAS/articles/GWAS.html"><strong>vignette</strong></a></p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<ul>
<li>Install from CRAN:</li>
</ul>
<!-- end list --><div class="sourceCode" id="cb1"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"statgenGWAS"</span>)</pre></div>
<ul>
<li>Install latest development version from GitHub (requires <a href="https://github.com/r-lib/remotes">remotes</a> package):</li>
</ul>
<!-- end list --><div class="sourceCode" id="cb2"><pre class="r"><span class="kw pkg">remotes</span><span class="kw ns">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span>(<span class="st">"Biometris/statgenGWAS"</span>, <span class="kw">ref</span> <span class="kw">=</span> <span class="st">"develop"</span>, <span class="kw">dependencies</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></div>
</div>
<div id="examples" class="section level2">
<h2 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h2>
<p>Example using the data from the European Union project DROPS. The same data is used in the vignette and described there in detail.</p>
<p>First get the data in the form required by the package, converting it to an object of class <strong>g</strong>(enomic)<strong>Data</strong>.</p>
<div class="sourceCode" id="cb3"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">statgenGWAS</span>)
<span class="co">## Read data.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"dropsMarkers"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"dropsMap"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"dropsPheno"</span>)

<span class="co">## Add genotypes as row names of dropsMarkers and drop Ind column.</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">dropsMarkers</span>) <span class="kw">&lt;-</span> <span class="no">dropsMarkers</span><span class="kw">[[</span><span class="st">"Ind"</span>]]
<span class="no">dropsMarkers</span> <span class="kw">&lt;-</span> <span class="no">dropsMarkers</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">dropsMarkers</span>) <span class="kw">!=</span> <span class="st">"Ind"</span>]

<span class="co">## Add genotypes as row names of dropsMap.</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">dropsMap</span>) <span class="kw">&lt;-</span> <span class="no">dropsMap</span><span class="kw">[[</span><span class="st">"SNP.names"</span>]]
<span class="co">## Rename Chomosome and Position columns.</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">dropsMap</span>)[<span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Chromosome"</span>, <span class="st">"Position"</span>), <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">dropsMap</span>))] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"chr"</span>, <span class="st">"pos"</span>)

<span class="co">## Rename Variety_ID in phenotypic data to genotype.</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">dropsPheno</span>)[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">dropsPheno</span>) <span class="kw">==</span> <span class="st">"Variety_ID"</span>] <span class="kw">&lt;-</span> <span class="st">"genotype"</span>
<span class="co">## Select relevant columns and convert data to a list.</span>
<span class="no">dropsPhenoList</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">dropsPheno</span>[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"genotype"</span>, <span class="st">"grain.yield"</span>,
                                         <span class="st">"grain.number"</span>, <span class="st">"seed.size"</span>,
                                         <span class="st">"anthesis"</span>, <span class="st">"silking"</span>, <span class="st">"plant.height"</span>,
                                         <span class="st">"tassel.height"</span>, <span class="st">"ear.height"</span>)],
                        <span class="kw">f</span> <span class="kw">=</span> <span class="no">dropsPheno</span><span class="kw">[[</span><span class="st">"Experiment"</span>]])

<span class="co">## Create a gData object all data.</span>
<span class="no">gDataDrops</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/gData.html">createGData</a></span>(<span class="kw">geno</span> <span class="kw">=</span> <span class="no">dropsMarkers</span>, <span class="kw">map</span> <span class="kw">=</span> <span class="no">dropsMap</span>, <span class="kw">pheno</span> <span class="kw">=</span> <span class="no">dropsPhenoList</span>)</pre></div>
<div class="sourceCode" id="cb4"><pre class="r"><span class="co">## Remove duplicate SNPs from gDataDrops.</span>
<span class="no">gDataDropsDedup</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/codeMarkers.html">codeMarkers</a></span>(<span class="no">gDataDrops</span>, <span class="kw">impute</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co">#&gt; Input contains 41722 SNPs for 246 genotypes.</span>
<span class="co">#&gt; 0 genotypes removed because proportion of missing values larger than or equal to 1.</span>
<span class="co">#&gt; 0 SNPs removed because proportion of missing values larger than or equal to 1.</span>
<span class="co">#&gt; 5098 duplicate SNPs removed.</span>
<span class="co">#&gt; Output contains 36624 SNPs for 246 genotypes.</span></pre></div>
<div class="sourceCode" id="cb5"><pre class="r"><span class="co">## Run single trait GWAS for traits 'grain.yield' and 'anthesis' for trial Mur13W.</span>
<span class="no">GWASDrops</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/runSingleTraitGwas.html">runSingleTraitGwas</a></span>(<span class="kw">gData</span> <span class="kw">=</span> <span class="no">gDataDropsDedup</span>,
                                <span class="kw">trials</span> <span class="kw">=</span> <span class="st">"Mur13W"</span>,
                                <span class="kw">traits</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"grain.yield"</span>, <span class="st">"anthesis"</span>))</pre></div>
<div class="sourceCode" id="cb6"><pre class="r"><span class="co">## QQ plot of GWAS Drops.</span>
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">GWASDrops</span>, <span class="kw">plotType</span> <span class="kw">=</span> <span class="st">"qq"</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="st">"grain.yield"</span>)</pre></div>
<p><img src="reference/figures/README-qqStg-1.png" width="75%"></p>
<div class="sourceCode" id="cb7"><pre class="r"><span class="co">## Manhattan plot of GWAS Drops.</span>
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">GWASDrops</span>, <span class="kw">plotType</span> <span class="kw">=</span> <span class="st">"manhattan"</span>, <span class="kw">trait</span> <span class="kw">=</span> <span class="st">"grain.yield"</span>)</pre></div>
<p><img src="reference/figures/README-manhattanStg-1.png" width="75%"></p>
<div class="sourceCode" id="cb8"><pre class="r"><span class="co">## Qtl plot of GWAS Drops.</span>
<span class="co">## Set significance threshold to 4 and normalize effect estimates.</span>
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">GWASDrops</span>, <span class="kw">plotType</span> <span class="kw">=</span> <span class="st">"qtl"</span>, <span class="kw">yThr</span> <span class="kw">=</span> <span class="fl">4</span>, <span class="kw">normalize</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></div>
<p><img src="reference/figures/README-qtlStgNorm-1.png" width="75%"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=statgenGWAS">https://â€‹cloud.r-project.org/â€‹package=statgenGWAS</a>
</li>
<li>Browse source code at <br><a href="https://github.com/Biometris/statgenGWAS/">https://â€‹github.com/â€‹Biometris/â€‹statgenGWAS/â€‹</a>
</li>
<li>Report a bug at <br><a href="https://github.com/Biometris/statgenGWAS/issues">https://â€‹github.com/â€‹Biometris/â€‹statgenGWAS/â€‹issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Bart-Jan van Rossum <br><small class="roles"> Author, maintainer </small>  </li>
<li>Willem Kruijer <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-7179-1733" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-pkg.org/pkg/statgenGWAS"><img src="https://www.r-pkg.org/badges/version/statgenGWAS"></a></li>
<li><a href="https://www.r-pkg.org/pkg/statgenGWAS"><img src="https://cranlogs.r-pkg.org/badges/statgenGWAS" alt="CRAN RStudio mirror downloads"></a></li>
<li><a href="https://github.com/Biometris/statgenGWAS/actions?workflow=R-CMD-check"><img src="https://github.com/Biometris/statgenGWAS/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://codecov.io/gh/Biometris/statgenGWAS"><img src="https://codecov.io/gh/Biometris/statgenGWAS/branch/master/graph/badge.svg" alt="codecov"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Bart-Jan van Rossum, Willem Kruijer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
