<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="statgenGWAS">
<title>Introduction to the statgenGWAS package • statgenGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to the statgenGWAS package">
<meta property="og:description" content="statgenGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">statgenGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.9</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/index.html">Vignettes</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    functions
  </a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown--other-packages">
    <span class="fa fa-seedling"></span>
     
    other packages
  </a>
  <div class="dropdown-menu" aria-labelledby="dropdown--other-packages">
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenSTA">statgenSTA</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenGxE">statgenGxE</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenHTP">statgenHTP</a>
  </div>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Biometris/statgenGWAS">
    <span class="fa fa-github fa-lg fab"></span>
     
    github
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="GWAS_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction to the statgenGWAS package</h1>
                        <h4 data-toc-skip class="author">Bart-Jan van Rossum</h4>
            
            <h4 data-toc-skip class="date">2022-10-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/Biometris/statgenGWAS/blob/HEAD/vignettes/GWAS.Rmd" class="external-link"><code>vignettes/GWAS.Rmd</code></a></small>
      <div class="d-none name"><code>GWAS.Rmd</code></div>
    </div>

    
    
<div class="section level2 unnumbered">
<h2 id="the-statgengwas-package">The statgenGWAS package<a class="anchor" aria-label="anchor" href="#the-statgengwas-package"></a>
</h2>
<p>The statgenGWAS package has been designed for performing single trait Genome Wide Association Studies (GWAS), where the objective is to find associations between genetic markers and phenotypic traits.</p>
<div class="section level3">
<h3 id="theoretical-background">Theoretical background<a class="anchor" aria-label="anchor" href="#theoretical-background"></a>
</h3>
<p>Suppose there are <span class="math inline">\(n\)</span> measurements of a phenotype for <span class="math inline">\(t\)</span> genotypes. Then a very common linear mixed model for association mapping is <span class="math display">\[y = X\beta + Zu + e,\]</span><br>
where <span class="math inline">\(y\)</span> is an <span class="math inline">\(n \times 1\)</span> vector of observed phenotypes and <span class="math inline">\(X\)</span> is an <span class="math inline">\(n \times q\)</span> design matrix for the fixed effects, including an intercept, covariates, and the SNP of interest. <span class="math inline">\(\beta\)</span> is the <span class="math inline">\(q \times 1\)</span> vector of coefficients of these fixed effects, the last element (<span class="math inline">\(\beta_{\textrm{SNP}}\)</span>) being the SNP-effect of interest. <span class="math inline">\(Z\)</span> is the <span class="math inline">\(n \times t\)</span> incidence matrix assigning individuals to genotypes. <span class="math inline">\(u\)</span> is the vector of genetic random effects, with <span class="math inline">\(var(u) = \sigma^2_g K\)</span>, where <span class="math inline">\(K\)</span> is a marker-based relatedness matrix. Finally, <span class="math inline">\(e\)</span> is a <span class="math inline">\(n \times 1\)</span> vector of residual (non-genetic) errors, with <span class="math inline">\(var(\epsilon) = \sigma^2_e I_n\)</span>. For this model the phenotypic variance-covariance matrix is <span class="math inline">\(\Sigma = \sigma^2_g ZKZ^{t} + \sigma^2_e I_n\)</span>.</p>
</div>
<div class="section level3">
<h3 id="application-in-the-package">Application in the package<a class="anchor" aria-label="anchor" href="#application-in-the-package"></a>
</h3>
<p>Single trait GWAS in the statgenGWAS package follows the approach of <span class="citation">Kang et al. (<a href="#ref-Kang2010" role="doc-biblioref">2010</a>)</span> and related papers:</p>
<ol style="list-style-type: decimal">
<li>For the ‘empty’ model (i.e., without any SNP in the fixed effects), we obtain REML-estimates of the variance components <span class="math inline">\(\sigma^2_g\)</span> and <span class="math inline">\(\sigma^2_e\)</span>, which give an estimate <span class="math inline">\(\hat \Sigma\)</span> of the variance-covariance matrix.</li>
<li>Fixing <span class="math inline">\(\hat \Sigma\)</span>, we use generalized least-squares (GLS) to test the null-hypothesis <span class="math inline">\(\beta_{\textrm{SNP}} = 0\)</span> using an F-test, for each SNP in turn. The last step gives vectors of estimated effect-size and P-values for all SNPs.</li>
</ol>
<p>These steps are repeated for each combination of trial and trait for which the analysis is run.</p>
<p>The REML-estimates in step (1) can be computed using either the Efficient Mixed Model Association (EMMA) algorithm <span class="citation">(Kang et al. <a href="#ref-Kang2008" role="doc-biblioref">2008</a>)</span>, or the Newton-Raphson (NR) algorithm <span class="citation">(Tunnicliffe <a href="#ref-Tunnicliffe1989" role="doc-biblioref">1989</a>)</span>. Our implementation of GLS is a c++ implementation of the R code of <span class="citation">Segura et al. (<a href="#ref-Segura2012" role="doc-biblioref">2012</a>)</span>.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="the-class-gdata">The class <code>gData</code><a class="anchor" aria-label="anchor" href="#the-class-gdata"></a>
</h2>
<p>Genotypic data are stored in an object of class <strong>gData</strong> (<strong>g</strong>enotypic <strong>Data</strong>). A gData object will contain all data needed for performing analyses, so the first thing to do when using the statgenGWAS package is creating a gData object. This can be done using the function <strong><em>createGData</em></strong>. A gData object combines genotypic data and phenotypic data in a <code>list</code>. In a typical workflow the user first imports marker and map data as illustrated in the following example.</p>
<p>If an existing gData object is provided as an input to the <code><a href="../reference/gData.html">createGData()</a></code> function, data is added to the existing object and existing data will be overwritten with a warning if necessary. All data is checked for consistency and, if needed, default names are added with a warning message.<br>
A gData object created this way will consist of the following components:</p>
<div class="section level3">
<h3 id="marker-map">Marker map<a class="anchor" aria-label="anchor" href="#marker-map"></a>
</h3>
<p>The data.frame <strong>map</strong> is used to describe the physical positions of the markers on the chromosomes. The data consists of two columns, <em>chr</em> for the name or number of the chromosome and <em>pos</em> for the position of the marker on the chromosome. The position can be in basepair or in centimorgan. The row names of map should be the names of the markers.</p>
</div>
<div class="section level3">
<h3 id="marker-matrix">Marker matrix<a class="anchor" aria-label="anchor" href="#marker-matrix"></a>
</h3>
<p>The marker matrix is stored in the matrix <strong>marker</strong> within the gData object. The column names should be the names of the markers and the row names the names of the genotypes. Markers can either be coded as character strings or as numerical values. In the former case, before performing any analysis, the marker matrix has to be converted to a numerical matrix. This can be do using the function <strong><em>codeMarkers</em></strong>.</p>
<p>enforce consistency between genotypic and phenotypic data.</p>
<div class="section level4">
<h4 id="phenotypic-data">Phenotypic data<a class="anchor" aria-label="anchor" href="#phenotypic-data"></a>
</h4>
<p>Phenotypic data, either directly from field trials or after summarizing can be stored in <strong>pheno</strong> in the gData object. Pheno can either be a single data.frame or a list of data.frames for storing data for different trials or different summarizations of the original data. The first column of all elements of pheno should be <em>genotype</em> and all the other columns should represent different traits. Storing additional variables should be done in <strong>covar</strong>. A single data.frame will be converted to a one-item list upon creation of a gData object.</p>
</div>
</div>
<div class="section level3">
<h3 id="kinship-matrix">Kinship matrix<a class="anchor" aria-label="anchor" href="#kinship-matrix"></a>
</h3>
<p>The kinship matrix describes the genetic relatedness between the different genotypes. It is stored in the matrix <strong>kinship</strong> in the gData object. This should be a square matrix with genotypes in both row and column names and a measure for the genetic relatedness in its cells. There are many ways of computing genetic relatedness some of which are included in this package. Instead of providing one kinship matrix a list of kinship matrices can be provided with each matrix in this list a chromosome specific kinship matrix, usually computed using all markers that are not on the corresponding chromosome. If a named list of kinship matrices is provided, its names should match the names or numbers of the chromosomes in <strong>map</strong>. Unnamed lists will be given default names.</p>
</div>
<div class="section level3">
<h3 id="further-covariates">Further covariates<a class="anchor" aria-label="anchor" href="#further-covariates"></a>
</h3>
<p>Further covariates that can be used in the analyses can be stored in the data.frame <strong>covar</strong>. This data.frame has genotypes in its row names and the covariates in the column names. Covariates can be numerical, factor or character data. The latter will be converted to a factor when creating the gData object.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="worked-example">Worked example<a class="anchor" aria-label="anchor" href="#worked-example"></a>
</h2>
<p>As an example of the functionality of the package a worked example is provided using maize data from the European Union project DROPS. The data is available from <a href="https://entrepot.recherche.data.gouv.fr/dataset.xhtml?persistentId=doi:10.15454/IASSTN" class="external-link uri">https://entrepot.recherche.data.gouv.fr/dataset.xhtml?persistentId=doi:10.15454/IASSTN</a> <span class="citation">(Millet et al. <a href="#ref-Millet2019" role="doc-biblioref">2019</a>)</span> and the relevant data sets are included as data.frames in the statgenGWAS package.</p>
<div class="section level3">
<h3 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h3>
<p>The first step is loading the data into R.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">dropsMarkers</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">dropsMap</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">dropsPheno</span><span class="op">)</span></span></code></pre></div>
<p>dropsMarkers contains the coded marker information for 41722 SNPs and 246 genotypes. dropsMap contains information about the positions of those SNPs on the B73 reference genome V2. dropsPheno contains data for the genotypic means (Best Linear Unbiased Estimators, BLUEs), with one value per experiment per genotype, for eight traits. For a more detailed description of the contents of the data see <code><a href="../reference/dropsData.html">help(dropsData)</a></code>.</p>
</div>
<div class="section level3">
<h3 id="create-gdata-object">Create gData object<a class="anchor" aria-label="anchor" href="#create-gdata-object"></a>
</h3>
<p>The next step is to create a <strong>gData</strong> object from the raw data that can be used for the GWAS analysis. For this the raw data has to be converted to a suitable format for a <strong>gData</strong> object. When creating a <strong>gData</strong> object, the markers should be input as a matrix or data.frame with genotypes in the row names and markers in the column names. The latter is already satisfied so only the row names should be added and the Ind column used for that should be dropped.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Add genotypes as row names of dropsMarkers and drop Ind column.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dropsMarkers</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">dropsMarkers</span><span class="op">[[</span><span class="st">"Ind"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">dropsMarkers</span> <span class="op">&lt;-</span> <span class="va">dropsMarkers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsMarkers</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"Ind"</span><span class="op">]</span></span></code></pre></div>
<p>The map used as input when creating a <strong>gData</strong> object, should at least have the columns “chr” and “pos” and the row names should be the names of the SNPs. Other columns are ignored when creating a <strong>gData</strong> object, so they can be left as they are.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Add genotypes as row names of dropsMap.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dropsMap</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">dropsMap</span><span class="op">[[</span><span class="st">"SNP.names"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## Rename Chomosome and Position columns.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsMap</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Chromosome"</span>, <span class="st">"Position"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsMap</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">"pos"</span><span class="op">)</span></span></code></pre></div>
<p>Now a gData object containing only map and marker information can be created. This object can be used as a base. Phenotypic data, a kinship matrix and covariates may be added later on.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create a gData object containing map and marker information.</span></span>
<span><span class="va">gDataDrops</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gData.html">createGData</a></span><span class="op">(</span>geno <span class="op">=</span> <span class="va">dropsMarkers</span>, map <span class="op">=</span> <span class="va">dropsMap</span><span class="op">)</span></span></code></pre></div>
<p>The next step involves adding phenotypic data to the gData object just created. dropsPheno contains genotypic means for 28 trials. To be able to run a GWAS analysis per trial the data has to be added as a list of 28 data.frames. These data.frames should have “genotype” as their first column and may only contain traits after that. Other columns need to be dropped.<br>
The phenotypic data can be added to <em>gDataDrops</em> using the same function that was used for creating the object. Adding an argument <code>gData</code> to the function call will add new data to an existing gData object. If gDataDrops would have been non-empty, the phenotypic data already present would be overwritten.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Rename Variety_ID to genotype.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsPheno</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsPheno</span><span class="op">)</span> <span class="op">==</span> <span class="st">"Variety_ID"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"genotype"</span></span>
<span><span class="co">## Select relevant columns and convert data to a list.</span></span>
<span><span class="va">dropsPhenoList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dropsPheno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"genotype"</span>, <span class="st">"grain.yield"</span>,</span>
<span>                                         <span class="st">"grain.number"</span>, <span class="st">"seed.size"</span>,</span>
<span>                                         <span class="st">"anthesis"</span>, <span class="st">"silking"</span>, <span class="st">"plant.height"</span>,</span>
<span>                                         <span class="st">"tassel.height"</span>, <span class="st">"ear.height"</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                        f <span class="op">=</span> <span class="va">dropsPheno</span><span class="op">[[</span><span class="st">"Experiment"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## Add phenotypic data to gDataDrops.</span></span>
<span><span class="va">gDataDrops</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gData.html">createGData</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDrops</span>, pheno <span class="op">=</span> <span class="va">dropsPhenoList</span><span class="op">)</span></span></code></pre></div>
<p>Note that if the phenotypic data consists of only one trial/experiment, it can be added as a single data.frame without first converting it to a list. In this case createGData will convert the input to a list with one item.</p>
<p>To get an idea of the contents of the data a summary of the gData object can be made. This will give an overview of the content of the map and markers and also print a summary per trait per trial. Since there are 28 trials and 8 traits in gDataDrops giving a very large summary, the output is restricted here to one trial, Mur13W, using the <code>trials</code> parameter of the summary function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Summarize gDataDrops.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gDataDrops</span>, trials <span class="op">=</span> <span class="st">"Mur13W"</span><span class="op">)</span></span>
<span><span class="co">#&gt; map</span></span>
<span><span class="co">#&gt;  Number of markers: 41722 </span></span>
<span><span class="co">#&gt;  Number of chromosomes: 10 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; markers</span></span>
<span><span class="co">#&gt;  Number of markers: 41722 </span></span>
<span><span class="co">#&gt;  Number of genotypes: 246 </span></span>
<span><span class="co">#&gt;  Content:</span></span>
<span><span class="co">#&gt;   0 1 2 NA </span></span>
<span><span class="co">#&gt;   0.28 0.01 0.71 0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; pheno</span></span>
<span><span class="co">#&gt;  Number of trials: 1 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Mur13W:</span></span>
<span><span class="co">#&gt;      Number of traits: 8 </span></span>
<span><span class="co">#&gt;      Number of genotypes: 246 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;         grain.yield grain.number seed.size anthesis silking plant.height tassel.height ear.height</span></span>
<span><span class="co">#&gt; Min.            3.3         1348       229       56      59          222            NA        102</span></span>
<span><span class="co">#&gt; 1st Qu.         6.3         2641       284       61      64          251            NA        125</span></span>
<span><span class="co">#&gt; Median          7.5         2965       298       63      66          258            NA        133</span></span>
<span><span class="co">#&gt; Mean            7.4         2986       297       63      66          259           NaN        133</span></span>
<span><span class="co">#&gt; 3rd Qu.         8.4         3359       312       66      68          266            NA        141</span></span>
<span><span class="co">#&gt; Max.           11.4         4510       365       71      74          294            NA        172</span></span>
<span><span class="co">#&gt; NA's            0.0            0         0        0       0            0           246          0</span></span></code></pre></div>
<p>It is also possible to plot the genetic map for a gData object. This will show the length of the chromosomes and the position of the markers in the gData object on the chromosomes. It is also possible to highlight one or more markers.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot genetic map.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gDataDrops</span><span class="op">)</span></span></code></pre></div>
<p><img src="GWAS_files/figure-html/plotGData-1.png" width="672"></p>
<p>To highlight a marker, the parameter <code>highlight</code> can be used. It requires <code>data.frame</code> with at least the column “chr” and “pos”. If a column name is present it will be used to annotate the highlighted marker. If no name column is present the annotation will be of the form pos@chr.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot genetic map.</span></span>
<span><span class="co">## Highlight the 20.000th marker in the map.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gDataDrops</span>, highlight <span class="op">=</span> <span class="va">dropsMap</span><span class="op">[</span><span class="fl">20000</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="GWAS_files/figure-html/plotGDataHL-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="recoding-and-cleaning-of-markers">Recoding and cleaning of markers<a class="anchor" aria-label="anchor" href="#recoding-and-cleaning-of-markers"></a>
</h3>
<p>Marker data has to be numerical and without missing values in order to do GWAS analysis. This can be achieved using the <em>codeMarkers</em> function. The marker data available for the DROPS project has already been converted from A/T/C/G to 0/1/2. The <em>codeMarkers</em> function in the package can still be used to further clean the markers, e.g. by removing the duplicate SNPs.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Remove duplicate SNPs from gDataDrops.</span></span>
<span><span class="va">gDataDropsDedup</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/codeMarkers.html">codeMarkers</a></span><span class="op">(</span><span class="va">gDataDrops</span>, impute <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Input contains 41722 SNPs for 246 genotypes.</span></span>
<span><span class="co">#&gt; 0 genotypes removed because proportion of missing values larger than or equal to 1.</span></span>
<span><span class="co">#&gt; 0 SNPs removed because proportion of missing values larger than or equal to 1.</span></span>
<span><span class="co">#&gt; 5098 duplicate SNPs removed.</span></span>
<span><span class="co">#&gt; Output contains 36624 SNPs for 246 genotypes.</span></span></code></pre></div>
<p>Note that duplicate SNPs are removed at random. To get reproducible results make sure to set a seed.</p>
<p>To demonstrate the options of the <em>codeMarkers</em> function, we randomly replace 1% of the values in the marker matrix by NA.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Copy gData object.</span></span>
<span><span class="va">gDataDropsMiss</span> <span class="op">&lt;-</span> <span class="va">gDataDrops</span></span>
<span><span class="co">## Add random missing values to 1% of the values in the marker matrix.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">nVal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gDataDropsMiss</span><span class="op">$</span><span class="va">markers</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">gDataDropsMiss</span><span class="op">$</span><span class="va">markers</span><span class="op">)</span></span>
<span><span class="va">gDataDropsMiss</span><span class="op">$</span><span class="va">markers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">nVal</span>, size <span class="op">=</span> <span class="va">nVal</span> <span class="op">/</span> <span class="fl">100</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span></code></pre></div>
<p>To re-impute the missing values added above, <em>codeMarkers</em> has to be run with <code>impute = TRUE</code>. Then the type of imputation can be chosen setting the parameter <code>imputeType</code>:</p>
<ul>
<li>
<code>imputeType = "fixed"</code>: Impute all missing values by a single fixed value. Use the parameter <code>fixedValue</code> to set this value.</li>
<li>
<code>imputeType = "random"</code>: Impute missing values with a random value based on the non-missing values for a SNP.</li>
<li>
<code>imputeType = "beagle"</code>: Impute missing values using the independent beagle software <span class="citation">(Browning and Browning <a href="#ref-Browning2007" role="doc-biblioref">2007</a>)</span>. Version 4.1 of this software is included in the package and is called with its default settings. It requires java to run. Since running beagle takes a long time for the current data, the call to beagle is only shown here but not run.</li>
</ul>
<p>Before imputation, genotypes with a high proportion of missing values in the marker matrix can be removed by setting the parameter nMissGeno. Setting it to 0.1 will remove all genotypes with a proportion of missing values higher than 0.1. In a similar way SNPs with a high proportion of missing values can be removed setting nMiss. The default value for both parameters is 1, so genotypes and SNPs with only missing values are always removed.<br>
The parameter MAF can be used to remove SNPs with a Minor Allele Frequency (MAF) lower than the specified MAF. As with removing duplicate SNPs this will be done both before and after imputation.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Impute missing values with random value.</span></span>
<span><span class="co">## Remove SNPs and genotypes with proportion of NA larger than 0.01.</span></span>
<span><span class="va">gDataDropsImputed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/codeMarkers.html">codeMarkers</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsMiss</span>,</span>
<span>                                 nMissGeno <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                                 nMiss <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                                 impute <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                 imputeType <span class="op">=</span> <span class="st">"random"</span>, </span>
<span>                                 verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Input contains 41722 SNPs for 246 genotypes.</span></span>
<span><span class="co">#&gt; 119 genotypes removed because proportion of missing values larger than or equal to 0.01.</span></span>
<span><span class="co">#&gt; 14341 SNPs removed because proportion of missing values larger than or equal to 0.01.</span></span>
<span><span class="co">#&gt; 882 duplicate SNPs removed.</span></span>
<span><span class="co">#&gt; 15213 missing values imputed.</span></span>
<span><span class="co">#&gt; 1046 duplicate SNPs removed after imputation.</span></span>
<span><span class="co">#&gt; Output contains 25453 SNPs for 127 genotypes.</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Impute missing values using beagle software.</span></span>
<span><span class="va">gDataDropsImputedBeagle</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/codeMarkers.html">codeMarkers</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsMiss</span>, </span>
<span>                                       impute <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                       imputeType <span class="op">=</span> <span class="st">"beagle"</span>,</span>
<span>                                       verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="single-trait-gwas">Single trait GWAS<a class="anchor" aria-label="anchor" href="#single-trait-gwas"></a>
</h3>
<p>The cleaned gData object can be used for performing single trait GWAS analysis with the <em>runSingleTraitGwas</em> function. The default parameter values in the function are chosen in such a way that usually the analysis gives decent results. In the examples below the trial Mur13W is used to demonstrate the options of the <em>runSingleTraitGwas</em> function.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run single trait GWAS for traits 'grain.yield' and 'anthesis' for trial Mur13W.</span></span>
<span><span class="va">GWASDrops</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSingleTraitGwas.html">runSingleTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grain.yield"</span>, <span class="st">"anthesis"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The output of the <em>runSingleTraitGwas</em> function is an object of class GWAS. This is a list consisting of five components described below.<br><strong>GWAResult</strong>: a list of data.tables, one for each trial for which the analysis was run. Each data.table has the following columns:</p>
<table class="table"><tbody>
<tr class="odd">
<td align="left">trait</td>
<td align="left">trait name</td>
</tr>
<tr class="even">
<td align="left">snp</td>
<td align="left">SNP name</td>
</tr>
<tr class="odd">
<td align="left">chr</td>
<td align="left">chromosome on which the SNP is located</td>
</tr>
<tr class="even">
<td align="left">pos</td>
<td align="left">position of the SNP on the chromosome</td>
</tr>
<tr class="odd">
<td align="left">allFreq</td>
<td align="left">allele frequency of the SNP</td>
</tr>
<tr class="even">
<td align="left">pValue</td>
<td align="left">P-value for the SNP</td>
</tr>
<tr class="odd">
<td align="left">effect</td>
<td align="left">effect of the SNP on the trait value</td>
</tr>
<tr class="even">
<td align="left">effectSe</td>
<td align="left">standard error of the effect of the SNP on the trait value</td>
</tr>
<tr class="odd">
<td align="left">RLR2</td>
<td align="left">likelihood-ratio-based R2 as defined in <span class="citation">Sun et al. (<a href="#ref-Sun2010" role="doc-biblioref">2010</a>)</span>
</td>
</tr>
<tr class="even">
<td align="left">LOD</td>
<td align="left">LOD score for the SNP, defined as <span class="math inline">\(-\log_{10}(pValue)\)</span>
</td>
</tr>
</tbody></table>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GWASDrops</span><span class="op">$</span><span class="va">GWAResult</span><span class="op">$</span><span class="va">Mur13W</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt;        trait           snp chr    pos allFreq pValue effect effectSe    RLR2   LOD</span></span>
<span><span class="co">#&gt;  grain.yield         SYN83   1   3498    0.60   0.38 -0.084    0.097 0.00310 0.417</span></span>
<span><span class="co">#&gt;  grain.yield PZE-101000060   1 157104    0.72   0.51  0.067    0.102 0.00176 0.291</span></span>
<span><span class="co">#&gt;  grain.yield PZE-101000088   1 238347    0.84   0.83 -0.027    0.125 0.00019 0.081</span></span>
<span><span class="co">#&gt;  grain.yield PZE-101000083   1 239225    0.58   0.22 -0.113    0.092 0.00615 0.660</span></span>
<span><span class="co">#&gt;  grain.yield PZE-101000108   1 255850    0.90   0.47 -0.124    0.170 0.00215 0.330</span></span>
<span><span class="co">#&gt;  grain.yield PZE-101000111   1 263938    0.83   0.32 -0.117    0.119 0.00395 0.489</span></span></code></pre></div>
<p>Note that the estimated effect is computed for a single allele. Its direction depends on the coding of the markers in the gData object. In this example the minor allele was used as reference allele, so the effects are the estimated effects for the minor allele.</p>
<p><strong>signSnp</strong>: a list of data.tables, one for each trial for which the analysis was run, containing the significant SNPs. Optionally also the SNPs close to the significant SNPs are included in the data.table. See <a href="#sign"><strong>3.6.7</strong></a> for more information on how to do this. The data.tables in signSnp consist of the same columns as those in GWAResult described above. Two extra columns are added:</p>
<table class="table">
<colgroup>
<col width="10%">
<col width="89%">
</colgroup>
<tbody>
<tr class="odd">
<td align="left">snpStatus</td>
<td align="left">either “significant SNP” or “within … of a significant SNP”</td>
</tr>
<tr class="even">
<td align="left">propSnpVar</td>
<td align="left">proportion of the variance explained by the SNP, computed as <span class="math inline">\(\beta_{\textrm{SNP}}^2 * var(\textrm{SNP}) / var(\textrm{pheno})\)</span>
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">GWASDrops</span><span class="op">$</span><span class="va">signSnp</span><span class="op">$</span><span class="va">Mur13W</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt;        trait           snp chr      pos allFreq  pValue effect effectSe  RLR2 LOD       snpStatus</span></span>
<span><span class="co">#&gt;  grain.yield PZE-106021363   6 18846283    0.70 3.5e-07   0.52      0.1 0.096 6.5 significant SNP</span></span>
<span><span class="co">#&gt;  grain.yield PZE-106021410   6 18990291    0.70 2.3e-07   0.53      0.1 0.098 6.6 significant SNP</span></span>
<span><span class="co">#&gt;  grain.yield PZE-106021419   6 18991091    0.74 2.9e-07   0.53      0.1 0.097 6.5 significant SNP</span></span>
<span><span class="co">#&gt;  grain.yield PZE-106021420   6 18991117    0.70 1.3e-06   0.50      0.1 0.087 5.9 significant SNP</span></span>
<span><span class="co">#&gt;  grain.yield PZE-106021424   6 18991481    0.74 5.0e-07   0.52      0.1 0.093 6.3 significant SNP</span></span>
<span><span class="co">#&gt;  propSnpVar</span></span>
<span><span class="co">#&gt;       0.088</span></span>
<span><span class="co">#&gt;       0.090</span></span>
<span><span class="co">#&gt;       0.083</span></span>
<span><span class="co">#&gt;       0.080</span></span>
<span><span class="co">#&gt;       0.080</span></span></code></pre></div>
<p><strong>kinship</strong>: the kinship matrix (or matrices) used in the GWAS analysis. This can either be the user provided kinship matrix or the kinship matrix computed when running the <em>runSingleTraitGwas</em> function.</p>
<p><strong>thr</strong>: a list of thresholds, one for each trial for which the analysis was run, used for determining significant SNPs.</p>
<p><strong>GWASInfo</strong>: additional information on the analysis, e.g. the call and the type of threshold used.</p>
</div>
<div class="section level3">
<h3 id="gwas-summary">GWAS Summary<a class="anchor" aria-label="anchor" href="#gwas-summary"></a>
</h3>
<p>For a quick overview of the results, e.g. the number of significant SNPs, use the summary function.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create summary of GWASDrops.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">GWASDrops</span><span class="op">)</span></span>
<span><span class="co">#&gt; Mur13W:</span></span>
<span><span class="co">#&gt;  Traits analysed: grain.yield, anthesis </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Data are available for 36624 SNPs.</span></span>
<span><span class="co">#&gt;   0 of them were not analyzed because their minor allele frequency is below 0.01 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  GLSMethod: single </span></span>
<span><span class="co">#&gt;  kinshipMethod: astle </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Trait: grain.yield </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      Mixed model with only polygenic effects, and no marker effects:</span></span>
<span><span class="co">#&gt;      Genetic variance: 0.96 </span></span>
<span><span class="co">#&gt;      Residual variance: 0.21 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      LOD-threshold: 5.9 </span></span>
<span><span class="co">#&gt;      Number of significant SNPs: 5 </span></span>
<span><span class="co">#&gt;      Smallest p-value among the significant SNPs: 2.3e-07 </span></span>
<span><span class="co">#&gt;      Largest p-value among the significant SNPs: 1.3e-06 (LOD-score: 5.9)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      No genomic control correction was applied</span></span>
<span><span class="co">#&gt;      Genomic control inflation-factor: 0.99 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Trait: anthesis </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      Mixed model with only polygenic effects, and no marker effects:</span></span>
<span><span class="co">#&gt;      Genetic variance: 4 </span></span>
<span><span class="co">#&gt;      Residual variance: 0.87 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      LOD-threshold: 5.9 </span></span>
<span><span class="co">#&gt;      Number of significant SNPs: 0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      No genomic control correction was applied</span></span>
<span><span class="co">#&gt;      Genomic control inflation-factor: 0.97</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="gwas-plots">GWAS Plots<a class="anchor" aria-label="anchor" href="#gwas-plots"></a>
</h3>
<p>The plot.GWAS function can be used to visualize the results in GWASDrops, with a QQ-plot, Manhattan plot or QTL-plot.</p>
<div class="section level4">
<h4 id="qq-plots">QQ-plots<a class="anchor" aria-label="anchor" href="#qq-plots"></a>
</h4>
<p>A QQ-plot of the observed against the expected <span class="math inline">\(-\log_{10}(p)\)</span> values can be made by setting <code>plotType = "qq"</code>. Most of the SNPs are expected to have no effect, resulting in P-values uniformly distributed on <span class="math inline">\([0,1]\)</span>, and leading to the identity function (<span class="math inline">\(y=x\)</span>) on the <span class="math inline">\(-\log_{10}(p)\)</span> scale. As in the plot below, deviations from this line should only occur on the right side of the plot, for a small number of SNPs with an effect on the phenotype (and possibly SNPs in LD). There is <strong>inflation</strong> if the observed <span class="math inline">\(-\log_{10}(p)\)</span> values are always above the line <span class="math inline">\(y=x\)</span>, and (less common) <strong>deflation</strong> if they are always below this line. A QQ-plot therefore gives a first impression of the quality of the GWAS model: if for example <span class="math inline">\(-\log_{10}(p)\)</span> values are consistently too large (inflation), the correction for genetic relatedness may not be adequate.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a QQ-plot of GWAS Drops.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"qq"</span>, trait <span class="op">=</span> <span class="st">"grain.yield"</span><span class="op">)</span></span></code></pre></div>
<p><img src="GWAS_files/figure-html/qqStg-1.png" width="672"></p>
</div>
<div class="section level4">
<h4 id="manhattan-plots">Manhattan plots<a class="anchor" aria-label="anchor" href="#manhattan-plots"></a>
</h4>
<p>A manhattan plot is made by setting <code>plotType = "manhattan"</code>. Significant SNPs are marked in red.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a manhattan plot of GWAS Drops.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"manhattan"</span>, trait <span class="op">=</span> <span class="st">"grain.yield"</span><span class="op">)</span></span></code></pre></div>
<p><img src="GWAS_files/figure-html/manhattanStg-1.png" width="672"></p>
<p>To change the threshold when plotting, without having to redo the GWAS analysis, use the parameter <code>yThr</code>. Use <code>chr</code> to plot a subset of chromosomes.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a manhattan plot of GWAS Drops.</span></span>
<span><span class="co">## Set significance threshold to 4 and only plot chromosomes 6 to 8.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"manhattan"</span>, trait <span class="op">=</span> <span class="st">"grain.yield"</span>, yThr <span class="op">=</span> <span class="fl">4</span>, chr <span class="op">=</span> <span class="fl">6</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span></span></code></pre></div>
<p><img src="GWAS_files/figure-html/manhattanStgThr-1.png" width="672"> When plotting a single chromosome it is possible to specify <code>startPos</code> and/or <code>endPos</code> to indicate the start position and end position for SNPs that are plotted.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a manhattan plot of GWAS Drops.</span></span>
<span><span class="co">## Set significance threshold to 4 and only plot first part of chromosome 6.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"manhattan"</span>, trait <span class="op">=</span> <span class="st">"grain.yield"</span>, </span>
<span>     yThr <span class="op">=</span> <span class="fl">4</span>, chr <span class="op">=</span> <span class="fl">6</span>, startPos <span class="op">=</span> <span class="fl">0</span>, endPos <span class="op">=</span> <span class="fl">6e7</span><span class="op">)</span></span></code></pre></div>
<p><img src="GWAS_files/figure-html/manhattanStgPos-1.png" width="672"></p>
<p>If the number of SNPs gets large, the plotting might get slow. This can be dealt with by setting the parameter <code>lod</code>. When doing so, for the SNPs with a LOD-value below <code>lod</code>, only 5% will be plotted. The chance of a SNP being plotted is proportional to its LOD-value. SNPs with a higher LOD have a higher chance of being plotted. Since most SNPs will have low LOD-values, this restriction doesn’t very much affect the resulting plot. Note that the 5% of the SNPs that is plotted, is selected randomly. For reproducible results make sure to set a seed.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a manhattan plot of GWAS Drops.</span></span>
<span><span class="co">## Plot only 5% of SNPs with a LOD below 3.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"manhattan"</span>, trait <span class="op">=</span> <span class="st">"grain.yield"</span>, lod <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="GWAS_files/figure-html/manhattanLod-1.png" width="672"></p>
<p>The user can specify known effects (<code>effects</code>), which will be used for determining true/false positives and false negatives. These are displayed in green, orange and yellow, respectively.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a manhattan plot of GWAS Drops with significance threshold 4.</span></span>
<span><span class="co">## Assume PZE-106021410 and PZE-105012420 are SNPs with known effects.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"manhattan"</span>, trait <span class="op">=</span> <span class="st">"grain.yield"</span>, </span>
<span>     effects <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PZE-106021410"</span>, <span class="st">"PZE-105012420"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GWAS_files/figure-html/manhattanEffects-1.png" width="672"> The plot now shows PZE-106021410 in green, since it is a SNP with a known effect also found by our GWAS analyses. The other SNPs on chromosome 6 are shown in orange, they are found by our GWAS, but not known effects. PZE-105012420 is shown in yellow, it is a SNP that has a known effect but it isn’t picked up as significant by our analysis. Note that PZE-106021410 and PZE-105012420 are chosen solely to demonstrate how the effects options works when making a manhattan plot, they don’t represent truly known effects.</p>
<p>Also the visual appearance of the plot can be altered, e.g by specifying the colors for the chromosomes or a custom title. See for specifics on how to do this <code><a href="../reference/plot.GWAS.html">help(plot.GWAS)</a></code>.</p>
</div>
<div class="section level4">
<h4 id="qtl-plots">QTL plots<a class="anchor" aria-label="anchor" href="#qtl-plots"></a>
</h4>
<p>A qtl plot can be made by setting <code>plotType = "qtl"</code>. In this plot the significant SNPs are marked by circles at their genomic positions, with diameter proportional to the estimated effect size; for an example see <span class="citation">Millet et al. (<a href="#ref-Millet2016" role="doc-biblioref">2016</a>)</span>. Typically, this is done for multiple traits or environments, with the genomic position on the x-axis, which are displayed horizontally above each other and can thus be compared.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a qtl plot of GWAS Drops for Mur13W.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"qtl"</span><span class="op">)</span></span></code></pre></div>
<p><img src="GWAS_files/figure-html/qtlStg-1.png" width="672"></p>
<p>In this case, for “anthesis” there are no significant SNPs so the line is left out. To change the threshold when plotting without having to redo the GWAS analysis, use the parameter <code>yThr</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a qtl plot of GWAS Drops for Mur13W.</span></span>
<span><span class="co">## Set significance threshold to 4.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"qtl"</span>, yThr <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="GWAS_files/figure-html/qtlStgThr-1.png" width="672"></p>
<p>Now there are significant SNPs for both traits. However since the traits are measured on a different scale, the effect estimates cannot be compared. For better comparison, one can set <code>normalize = TRUE</code>, which divides the estimates by the standard deviation of the phenotype.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a qtl plot of GWAS Drops for Mur13W.</span></span>
<span><span class="co">## Set significance threshold to 4 and normalize effect estimates.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"qtl"</span>, yThr <span class="op">=</span> <span class="fl">4</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="GWAS_files/figure-html/qtlStgNorm-1.png" width="672"></p>
<p>Now the effects are on the same scale and comparable.</p>
<p>Other parameters can be used to plot a subset of the chromosomes (<code>chr</code>) and directly export the plot to .pptx (<code>exportPptx = TRUE</code> and specify <code>pptxName</code>). Note that the officer package is required for this. A full list of parameters can be found by running <code><a href="../reference/plot.GWAS.html">help(plot.GWAS)</a></code>.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="kinship-matrices">Kinship matrices<a class="anchor" aria-label="anchor" href="#kinship-matrices"></a>
</h2>
<p>The <em>runSingleTraitGwas</em> function has an argument <code>kinshipMethod</code>, which defines the kinship matrix used for association mapping. Kinship matrices can be computed directly using the <em>kinship</em> function or within the <em>runSingleTraitGwas</em> function. There are four options:</p>
<ol style="list-style-type: decimal">
<li>using the covariance between the scaled SNP-scores (<code>kinshipMethod = "astle"</code>, the default; see e.g. equation (2.2) in <span class="citation">Astle and Balding (<a href="#ref-Astle2009" role="doc-biblioref">2009</a>)</span>)</li>
<li>Identity by State (<code>kinshipMethod = "IBS"</code>; see e.g. equation (2.3) in <span class="citation">Astle and Balding (<a href="#ref-Astle2009" role="doc-biblioref">2009</a>)</span>)</li>
<li>using the formula by <span class="citation">VanRaden (<a href="#ref-VanRaden2008" role="doc-biblioref">2008</a>)</span> (<code>kinshipMethod = "vanRaden"</code>)</li>
<li>User-defined, in which case the parameter <code>kin</code> needs to be specified.</li>
</ol>
<p>By default, the same kinship matrix is used for testing all SNPs (<code>GLSMethod = "single"</code>). When <code>GLSMethod = "multi"</code>, the kinship matrix is chromosome-specific. In this case, the function fits variance components and computes effect-estimates and P-values for each chromosome in turn, using the kinship matrix for that chromosome. Each chromosome-specific kinship matrix is computed using the method specified by the argument <code>kinshipMethod</code>, using all SNPs that are <strong>not</strong> on this chromosome. As shown by <span class="citation">Rincent et al. (<a href="#ref-Rincent2014" role="doc-biblioref">2014</a>)</span>, this often gives a considerable improvement in power.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run single trait GWAS for trial 'Mur13W' and trait 'grain.yield'</span></span>
<span><span class="co">## Use chromosome specific kinship matrices computed using method of van Raden.</span></span>
<span><span class="va">GWASDropsChrSpec</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSingleTraitGwas.html">runSingleTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>, </span>
<span>                                       traits <span class="op">=</span> <span class="st">"grain.yield"</span>,</span>
<span>                                       trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                       GLSMethod <span class="op">=</span> <span class="st">"multi"</span>,</span>
<span>                                       kinshipMethod <span class="op">=</span> <span class="st">"vanRaden"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="opts">Further options<a class="anchor" aria-label="anchor" href="#opts"></a>
</h2>
<p>The <em>runSingleTraitGwas</em> function has many more parameters that can be specified. In this section similar parameters are grouped and explained with examples on how to use them.</p>
<div class="section level3">
<h3 id="multiple-trials-and-traits">Multiple trials and traits<a class="anchor" aria-label="anchor" href="#multiple-trials-and-traits"></a>
</h3>
<p>The analysis can be performed for more than one trait in a single function call. Not specifying <code>traits</code> at all will perform the analysis for all traits that are present in the phenotypic data in the gData object. The results will be combined in a single output object with the traits specified in the column <em>trait</em> in <em>GWAResult</em> and <em>signSnp</em> within this object. Running for multiple trials is also possible and will generate a list of outputs within <em>GWAResult</em> and <em>signSnp</em> in the output object. Not specifying <code>trials</code> will run the GWAS analysis for all trials in the phenotypic data in the gData object.</p>
</div>
<div class="section level3">
<h3 id="significance-thresholds">Significance thresholds<a class="anchor" aria-label="anchor" href="#significance-thresholds"></a>
</h3>
<p>The threshold for selecting significant SNPs in a GWAS analysis is computed by default using Bonferroni correction, with an alpha of 0.05. The alpha can be modified setting the option <em>alpha</em> when calling <em>runSingleTraitGwas</em>. Two other threshold types can be used: a fixed threshold (<code>thrType = "fixed"</code>) specifying the <span class="math inline">\(-\log_{10}(p)\)</span> (<em>LODThr</em>) value of the threshold, or a threshold that defines the n SNPs with the highest <span class="math inline">\(-\log_{10}(p)\)</span> scores as significant SNPs. Set <code>thrType = "small"</code> together with <code>nSnpLOD = n</code> to do this. In the following example, we select all SNPs with <span class="math inline">\(p &lt; 10^{-4}\)</span>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run single trait GWAS for trait 'grain.yield' for Mur13W.</span></span>
<span><span class="co">## Use a fixed significance threshold of 4.</span></span>
<span><span class="va">GWASDropsFixThr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSingleTraitGwas.html">runSingleTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                      trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                      traits <span class="op">=</span> <span class="st">"grain.yield"</span>,</span>
<span>                                      thrType <span class="op">=</span> <span class="st">"fixed"</span>,</span>
<span>                                      LODThr <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="controlling-false-discovery-rate">Controlling false discovery rate<a class="anchor" aria-label="anchor" href="#controlling-false-discovery-rate"></a>
</h4>
<p>A final option for selecting significant SNPs is by setting <code>thrType = "fdr"</code>. When doing so the significant SNPs won’t be selected by computing a genome wide threshold, but by trying to control the rate of false discoveries as in <span class="citation">Brzyski et al. (<a href="#ref-Brzyski2016" role="doc-biblioref">2016</a>)</span>.</p>
<p>First, a list is defined containing all SNPs with a p-Value below <code>pThr</code>. Then clusters of SNPs are created using a two step iterative process in which SNPs with the lowest p-values are selected as cluster representatives. This SNP and all SNPs that have a correlation with this SNP of <span class="math inline">\(\rho\)</span> or higher (specified by the function parameter <code>rho</code>) will form a cluster. The selected SNPs are removed from the list and the procedure is repeated until no SNPs are left. At the end of this step, one has a list of clusters, with corresponding vector of p-values of the cluster representatives. Finally, to determine the number of significant clusters, the first cluster is determined for which the p-value of the cluster representative is larger than <span class="math inline">\(cluster_{number} * \alpha / m\)</span>, where <span class="math inline">\(m\)</span> is the number of SNPs and <span class="math inline">\(\alpha\)</span> can be specified by the corresponding function parameter. All previous clusters are selected as significant.</p>
<p>Note that because of the way the significant SNPs are selected it is possible that a SNP with a lower p-Value than a significant SNP is not significant itself. This happens if that SNP is highly correlated to another SNP with an even lower p-Value. Because of this it is not possible to determine a threshold for the p-Value below which SNPs are significant. The value for the threshold in the output will therefor be <code>NA</code> and no line will be plotted for it in the manhattan plot.</p>
</div>
</div>
<div class="section level3">
<h3 id="variance-covariance-matrix">Variance covariance matrix<a class="anchor" aria-label="anchor" href="#variance-covariance-matrix"></a>
</h3>
<p>There are two ways to compute the phenotypic variance covariance matrix used in the GWAS analysis. Either the EMMA algorithm or the Newton-Raphson algorithm. Specify the method by setting the parameter <code>remlAlgo</code> to either <code>"EMMA"</code> or <code>"NR"</code>. By default the EMMA algorithm is used.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run single trait GWAS for trait 'grain.yield' for Mur13W.</span></span>
<span><span class="co">## Use the Newton Raphson algorithm for computing the variance components.</span></span>
<span><span class="va">GWASDropsNR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSingleTraitGwas.html">runSingleTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                  trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                  traits <span class="op">=</span> <span class="st">"grain.yield"</span>,</span>
<span>                                  remlAlgo <span class="op">=</span> <span class="st">"NR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="gcCorr">Genomic control correction<a class="anchor" aria-label="anchor" href="#gcCorr"></a>
</h3>
<p>In the GWAS object returned by <em>runSingleTraitGwas</em>, the inflation factor <span class="citation">(Devlin and Roeder <a href="#ref-Devlin1999" role="doc-biblioref">1999</a>)</span> is returned in the GWASInfo. It is also shown in the summary. Ideally this factor should be 1, meaning there is no inflation at all. In practice however this rarely happens. Its value for Mur13W can be checked as follows:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GWASDrops</span><span class="op">$</span><span class="va">GWASInfo</span><span class="op">$</span><span class="va">inflationFactor</span><span class="op">$</span><span class="va">Mur13W</span></span>
<span><span class="co">#&gt; grain.yield    anthesis </span></span>
<span><span class="co">#&gt;        0.99        0.97</span></span></code></pre></div>
<p>Both values are very close to 1, so in this case there is no need to apply any correction based on this. If the values are further away from 1, the inflation can be corrected for by setting <code>genomicControl = TRUE</code> in <em>runSingleTraitGwas</em>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run single trait GWAS for trait 'grain.yield' for Mur13W.</span></span>
<span><span class="co">## Perform genomic correction on the p-Values.</span></span>
<span><span class="va">GWASDropsGenControl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSingleTraitGwas.html">runSingleTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                          trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                          traits <span class="op">=</span> <span class="st">"grain.yield"</span>,</span>
<span>                                          genomicControl <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="covariates">Covariates<a class="anchor" aria-label="anchor" href="#covariates"></a>
</h3>
<p>Covariates can be included as extra fixed effects in the GWAS model. The <em>runSingleTraitGwas</em> function distinguishes between ‘usual’ covariates and SNP-covariates. The former could be design factors such as block, or other traits one wants to condition on. In the latter case, the covariate(s) are one or more of the markers contained in the genotypic data. SNP-covariates can be set with the parameter <code>snpCov</code>, which should be a vector of marker-names. Similarly, other covariates should be specified using the parameter <code>covar</code>, containing a vector of covariate names. The gData object should contain these covariates in gData$covar.</p>
<p>In case SNP-covariates are used, GWAS for all the other SNPs is performed with the the SNP-covariates as extra fixed effect; also the null model used to estimate the variance components includes these effects. For each SNP in SNP-covariates, a p-value is obtained using the same F-test and null model to estimate the variance components, but with only all other SNPs (if any) in SNP-covariates as fixed effects.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run single trait GWAS for trait 'grain.yield' for Mur13W.</span></span>
<span><span class="co">## Use PZE-106021410, the most significant SNP, a SNP covariate.</span></span>
<span><span class="va">GWASDropsSnpCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSingleTraitGwas.html">runSingleTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                      trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                      traits <span class="op">=</span> <span class="st">"grain.yield"</span>,</span>
<span>                                      snpCov <span class="op">=</span> <span class="st">"PZE-106021410"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="minor-allele-frequency">Minor Allele Frequency<a class="anchor" aria-label="anchor" href="#minor-allele-frequency"></a>
</h3>
<p>It is recommended to remove SNPs with a low minor allele frequency (MAF) from the data before starting a GWAS analysis. However it is also possible to do so in the analysis itself. The difference between these approaches is that <em>codeMarkers</em> removes the SNPs, whereas <em>runSingleTraitGwas</em> excludes them from the analysis but leaves them in the output (with results set to NA). In the latter case it will still be possible to see the allele frequency of the SNP.<br>
By default all SNPs with a MAF lower than 0.01 are excluded from the analysis. This can be controlled by the parameter <code>MAF</code>. Setting MAF to 0 will still exclude duplicate SNPs since duplicates cause problems when fitting the underlying models. Instead of excluding SNPs with a low MAF it is also possible to exclude SNPs with a low minor allele count (MAC). This is done setting <code>useMAF = FALSE</code> and then specifying <code>MAC</code>.<br>
Note that the computation of allele frequencies in <em>runSingleTraitGwas</em> is done after the removal of genotypes that have missing values for the trait for which the analysis is done. This could lead to slightly different values for the allele frequencies than when these are computed in <em>codeMarkers</em> (which only uses the genotypic data) resulting in the removal a slightly different set of markers.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run single trait GWAS for trait 'grain.yield' for Mur13W.</span></span>
<span><span class="co">## Only include SNPs that have a MAC of at least 20</span></span>
<span><span class="va">GWASDropsMAC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSingleTraitGwas.html">runSingleTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                   trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                   traits <span class="op">=</span> <span class="st">"grain.yield"</span>,</span>
<span>                                   useMAF <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                   MAC <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sign">SNPs close to significant SNPs<a class="anchor" aria-label="anchor" href="#sign"></a>
</h3>
<p>Apart from the significant SNPs themselves, it is usually of interest to identify SNPs that are close to the significant SNPs. These can be included in the output by setting <code>sizeIncludedRegion</code> to a positive value. When doing so, SNPs that are within a distance of <code>sizeIncludedRegion</code> from a significant SNP (based on the information in the map) are included in the signSnp data.table in the output. Only SNPs having sufficient linkage disequilibrium (LD) with the significant SNP are included. LD is measured in terms of the squared Pearson correlation (<span class="math inline">\(R^2\)</span>). The minimum <span class="math inline">\(R^2\)</span> required to be included can be specified by setting <code>minR2</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run single trait GWAS for trait 'grain.yield' for Mur13W.</span></span>
<span><span class="co">## Include SNPs within 200000 centimorgan of significant SNPs with a minimum LD of 0.1.</span></span>
<span><span class="va">GWASDropsInclClose</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSingleTraitGwas.html">runSingleTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                         trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                         traits <span class="op">=</span> <span class="st">"grain.yield"</span>,</span>
<span>                                         sizeInclRegion <span class="op">=</span> <span class="fl">200000</span>,</span>
<span>                                         minR2 <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="co">## Check signSnp in output.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GWASDropsInclClose</span><span class="op">$</span><span class="va">signSnp</span><span class="op">$</span><span class="va">Mur13W</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt;        trait           snp chr      pos allFreq  pValue effect effectSe   RLR2  LOD</span></span>
<span><span class="co">#&gt;  grain.yield PZE-106021250   6 18528809    0.88 8.4e-02 -0.260     0.15 0.0121 1.08</span></span>
<span><span class="co">#&gt;  grain.yield PZE-106021260   6 18535896    0.84 5.5e-01 -0.080     0.13 0.0015 0.26</span></span>
<span><span class="co">#&gt;  grain.yield PZE-106021293   6 18645872    0.94 5.5e-01  0.110     0.18 0.0015 0.26</span></span>
<span><span class="co">#&gt;  grain.yield PZE-106021308   6 18646399    0.84 5.3e-01 -0.083     0.13 0.0016 0.27</span></span>
<span><span class="co">#&gt;  grain.yield PZE-106021363   6 18846283    0.70 3.5e-07  0.522     0.10 0.0959 6.46</span></span>
<span><span class="co">#&gt;  grain.yield PZE-106021410   6 18990291    0.70 2.3e-07  0.527     0.10 0.0984 6.63</span></span>
<span><span class="co">#&gt;                          snpStatus propSnpVar</span></span>
<span><span class="co">#&gt;  within 2e+05 of a significant SNP     0.0115</span></span>
<span><span class="co">#&gt;  within 2e+05 of a significant SNP     0.0013</span></span>
<span><span class="co">#&gt;  within 2e+05 of a significant SNP     0.0010</span></span>
<span><span class="co">#&gt;  within 2e+05 of a significant SNP     0.0015</span></span>
<span><span class="co">#&gt;                    significant SNP     0.0884</span></span>
<span><span class="co">#&gt;                    significant SNP     0.0903</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2 unnumbered">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references">
<div id="ref-Astle2009">
<p>Astle, William, and David J. Balding. 2009. “Population Structure and Cryptic Relatedness in Genetic Association Studies.” <em>Statistical Science</em> 24 (4): 451–71. <a href="https://doi.org/10.1214/09-sts307" class="external-link">https://doi.org/10.1214/09-sts307</a>.</p>
</div>
<div id="ref-Browning2007">
<p>Browning, Sharon R., and Brian L. Browning. 2007. “Rapid and Accurate Haplotype Phasing and Missing-Data Inference for Whole-Genome Association Studies by Use of Localized Haplotype Clustering.” <em>The American Journal of Human Genetics</em> 81 (5): 1084–97. <a href="https://doi.org/10.1086/521987" class="external-link">https://doi.org/10.1086/521987</a>.</p>
</div>
<div id="ref-Brzyski2016">
<p>Brzyski, Damian, Christine B. Peterson, Piotr Sobczyk, Emmanuel J. Candès, Malgorzata Bogdan, and Chiara Sabatti. 2016. “Controlling the Rate of GWAS False Discoveries.” <em>Genetics</em> 205 (1): 61–75. <a href="https://doi.org/10.1534/genetics.116.193987" class="external-link">https://doi.org/10.1534/genetics.116.193987</a>.</p>
</div>
<div id="ref-Devlin1999">
<p>Devlin, B., and Kathryn Roeder. 1999. “Genomic Control for Association Studies.” <em>Biometrics</em> 55 (4): 997–1004. <a href="https://doi.org/10.1111/j.0006-341x.1999.00997.x" class="external-link">https://doi.org/10.1111/j.0006-341x.1999.00997.x</a>.</p>
</div>
<div id="ref-Kang2010">
<p>Kang, Hyun Min, Jae Hoon Sul, Susan K Service, Noah A Zaitlen, Sit-yee Kong, Nelson B Freimer, Chiara Sabatti, and Eleazar Eskin. 2010. “Variance Component Model to Account for Sample Structure in Genome-Wide Association Studies.” <em>Nature Genetics</em> 42 (4): 348–54. <a href="https://doi.org/10.1038/ng.548" class="external-link">https://doi.org/10.1038/ng.548</a>.</p>
</div>
<div id="ref-Kang2008">
<p>Kang, Hyun Min, Noah A. Zaitlen, Claire M. Wade, Andrew Kirby, David Heckerman, Mark J. Daly, and Eleazar Eskin. 2008. “Efficient Control of Population Structure in Model Organism Association Mapping.” <em>Genetics</em> 178 (3): 1709–23. <a href="https://doi.org/10.1534/genetics.107.080101" class="external-link">https://doi.org/10.1534/genetics.107.080101</a>.</p>
</div>
<div id="ref-Millet2019">
<p>Millet, Emilie J., Cyril Pommier, Mélanie Buy, Axel Nagel, Willem Kruijer, Therese Welz-Bolduan, Jeremy Lopez, et al. 2019. “A Multi-Site Experiment in a Network of European Fields for Assessing the Maize Yield Response to Environmental Scenarios.” Portail Data Inra. <a href="https://doi.org/10.15454/IASSTN" class="external-link">https://doi.org/10.15454/IASSTN</a>.</p>
</div>
<div id="ref-Millet2016">
<p>Millet, Emilie, Claude Welcker, Willem Kruijer, Sandra Negro, Stephane Nicolas, Sebastien Praud, Nicolas Ranc, et al. 2016. “Genome-Wide Analysis of Yield in Europe: Allelic Effects as Functions of Drought and Heat Scenarios.” <em>Plant Physiology</em>, July, pp.00621.2016. <a href="https://doi.org/10.1104/pp.16.00621" class="external-link">https://doi.org/10.1104/pp.16.00621</a>.</p>
</div>
<div id="ref-Rincent2014">
<p>Rincent, Renaud, Laurence Moreau, Hervé Monod, Estelle Kuhn, Albrecht E. Melchinger, Rosa A. Malvar, Jesus Moreno-Gonzalez, et al. 2014. “Recovering Power in Association Mapping Panels with Variable Levels of Linkage Disequilibrium.” <em>Genetics</em> 197 (1): 375–87. <a href="https://doi.org/10.1534/genetics.113.159731" class="external-link">https://doi.org/10.1534/genetics.113.159731</a>.</p>
</div>
<div id="ref-Segura2012">
<p>Segura, Vincent, Bjarni J Vilhjálmsson, Alexander Platt, Arthur Korte, Ümit Seren, Quan Long, and Magnus Nordborg. 2012. “An Efficient Multi-Locus Mixed-Model Approach for Genome-Wide Association Studies in Structured Populations.” <em>Nature Genetics</em> 44 (7): 825–30. <a href="https://doi.org/10.1038/ng.2314" class="external-link">https://doi.org/10.1038/ng.2314</a>.</p>
</div>
<div id="ref-Sun2010">
<p>Sun, G, C Zhu, M H Kramer, S-S Yang, W Song, H-P Piepho, and J Yu. 2010. “Variation Explained in Mixed-Model Association Mapping.” <em>Heredity</em> 105 (4): 333–40. <a href="https://doi.org/10.1038/hdy.2010.11" class="external-link">https://doi.org/10.1038/hdy.2010.11</a>.</p>
</div>
<div id="ref-Tunnicliffe1989">
<p>Tunnicliffe, G Wilson. 1989. “On the Use of Marginal Likelihood in Time Series Model Estimation.” <em>JRSS</em> 51 (1): 15–27.</p>
</div>
<div id="ref-VanRaden2008">
<p>VanRaden, P. M. 2008. “Efficient Methods to Compute Genomic Predictions.” <em>Journal of Dairy Science</em> 91 (11): 4414–23. <a href="https://doi.org/10.3168/jds.2007-0980" class="external-link">https://doi.org/10.3168/jds.2007-0980</a>.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Bart-Jan van Rossum, Willem Kruijer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
